{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <p style=\"text-align: right;\"> &#9989; **Group 1 (GoGreen)** </p>\n",
    "#### <p style=\"text-align: right;\"> &#9989; Beth, Zhongjie, McKenna, Erik</p>\n",
    "\n",
    "# Module 1\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How do SNPs relate to gene transcription?\n",
    "\n",
    "We will work with maize. Last updated 2020/01/23"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Jupyter by default converts `tabs` into `spaces` which is good."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, import the usual libraries\n",
    "- `math`: basic math operations\n",
    "- `os`: enable file manipulation with the OS\n",
    "- `sys`: enable interaction with commandline\n",
    "- `glob`: more variable manipulation\n",
    "- `matplotlib.pyplot`: default plotter (I personally like ggplot waaaaay better. E)\n",
    "    - `inline`: so that plots are shown in the notebook\n",
    "- `seaborn`: nicer plots\n",
    "- `numpy`: all number cruching done here\n",
    "- `pandas`: data wrangling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math  \n",
    "import os   \n",
    "import glob \n",
    "import matplotlib.pyplot as plt \n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define some functions. Let's see for how long can we keep tidy commented code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dot_to_int(char_in, substitute='.', value=0.0):\n",
    "    \"\"\"\n",
    "    Substitute a string for a given value if such string is of interest\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    char_in : string\n",
    "         possible char to be substituted\n",
    "    \n",
    "    substitute : string\n",
    "         if the given input matches this object, substitute\n",
    "    \n",
    "    value : int\n",
    "         if applicable, substitute the input with this value\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    value : if input matches substitute, else returns unchanged input\n",
    "    \"\"\"\n",
    "    if(char_in == substitute):\n",
    "        return value\n",
    "    else:\n",
    "        return char_in\n",
    "    \n",
    "def numMissing(row, start=10, missing='.'):\n",
    "    foo = row[start:] == missing\n",
    "    return(foo.sum())\n",
    "\n",
    "\n",
    "def numData(row, data_cols=['Total Calls', 'Missing Data']):\n",
    "    return row[data_cols].sum()  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_Sidenote:_ We can print whatever text was written between the `\"\"\"` above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    Substitute a string for a given value if such string is of interest\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    char_in : string\n",
      "         possible char to be substituted\n",
      "    \n",
      "    substitute : string\n",
      "         if the given input matches this object, substitute\n",
      "    \n",
      "    value : int\n",
      "         if applicable, substitute the input with this value\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    value : if input matches substitute, else returns unchanged input\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "print(dot_to_int.__doc__)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make sure to write the right path to the files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to the folder where all the raw data is located\n",
    "src = '/home/ejam/documents/css893/raw_data/'\n",
    "\n",
    "# Path where we will save outputs\n",
    "dst = '/home/ejam/documents/css893/output/'\n",
    "\n",
    "# Matrix with raw SNP data\n",
    "file_snp_raw = 'B73_plus_RTAs_snp_matrix_995785.txt'\n",
    "\n",
    "# Matrix with imputed SNP data\n",
    "file_snp_imp = 'B73_plus_RTAs_snp_matrix_imputed/widiv_942g_979873SNPs_imputed_filteredGenos_withRTA_AGPv4.hmp.txt'\n",
    "\n",
    "# Matrix with gene expression (FPKM)\n",
    "file_gen = '942_FPKM_B73_genes_w_feature.txt'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the SNP matrix. __Loading the full SNP matrix consumes about 16Gb!!!__ You may need more to do some basic data wrangling. I need way more memory to do fancier stuff ðŸ˜¿.\n",
    "If you don't have at least 32GB of RAM in your machine, you can simply load the first 1000 SNPs. This will give you a sense of how the data looks like."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# BEWARE : Line below consumes 16Gb of RAM at some point. \n",
    "# Trying to do some basic wrangling (i.e. masking) may consume additional 8Gb\n",
    "# More sophisticated wrangling (apply the same function to all rows) may consume much more RAM\n",
    "\n",
    "# We already know that beyond SNP 899784, the information is irrelevant (contigs and RTAs)\n",
    "# Set clean = False to see how such cutoff was computed\n",
    "clean = True\n",
    "\n",
    "if clean:\n",
    "    snp_matrix_imputed = pd.read_table(src + file_snp_imp, dtype={'chrom': int}, nrows=899784)\n",
    "else:\n",
    "    snp_matrix_imputed = pd.read_table(src + file_snp_imp, dtype={'chrom': int})\n",
    "\n",
    "# Run the commented line below instead if you don't have much RAM\n",
    "#snp_matrix_imputed = pd.read_table(src + file_snp, dtype={'chromosome': str}, nrows=1000)\n",
    "snp_matrix_raw = pd.read_table(src + file_snp_raw, dtype={'chromosome': str}, nrows=1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get a sense of how the data is arranged in the raw SNP matrix. The full matrix has 995785 rows Ã— 952 columns.\n",
    "- Each row is a SNP\n",
    "- The first ten columns are metadata\n",
    "- The other 942 correspond to different cultivars.\n",
    "\n",
    "The column names for cultivars is found in `B73_plus_RTAs_snp_matrix_imputed/widiv_942g_979873SNPs_imputed_filteredGenos_withRTA_AGPv4.hmp.txt`\n",
    "\n",
    "You may notice that `Missing Data` + `Total Calls` = 959 for every SNP. However, we only have information from 942 cultivars. That is, the data suggests there are 17 samples are missing. From Gage _et al_ 2019:\n",
    "\n",
    "> Based on their location in the tree and known pedigree \n",
    "information, the inbred line Mo8W was removed from the \n",
    "SNP and fragments per kb of exon model per million mapped\n",
    "reads (FPKM) matrix files. An additional 16 inbred lines\n",
    "were also removed from the SNP and FPKM files, resulting\n",
    "in a final set of 942 inbreds, referred to as the \n",
    "WiDiv-942 (Mazaheri et al.,2019), which were analyzed in \n",
    "this study."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>Allele Counts</th>\n",
       "      <th>Missing Data</th>\n",
       "      <th>Total Calls</th>\n",
       "      <th>Percent A</th>\n",
       "      <th>Percent T</th>\n",
       "      <th>Percent C</th>\n",
       "      <th>Percent G</th>\n",
       "      <th>...</th>\n",
       "      <th>NC328</th>\n",
       "      <th>NC326</th>\n",
       "      <th>PHV53</th>\n",
       "      <th>DKIBC2</th>\n",
       "      <th>A641</th>\n",
       "      <th>WIL900</th>\n",
       "      <th>Va22</th>\n",
       "      <th>E8501</th>\n",
       "      <th>PHP85</th>\n",
       "      <th>Oh43</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>B73V4_ctg102</td>\n",
       "      <td>36353</td>\n",
       "      <td>A</td>\n",
       "      <td>A(0)T(0)C(0)G(264)</td>\n",
       "      <td>695</td>\n",
       "      <td>264</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>100</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B73V4_ctg10</td>\n",
       "      <td>292327</td>\n",
       "      <td>T</td>\n",
       "      <td>A(0)T(744)C(0)G(105)</td>\n",
       "      <td>110</td>\n",
       "      <td>849</td>\n",
       "      <td>.</td>\n",
       "      <td>87.6325088339223</td>\n",
       "      <td>.</td>\n",
       "      <td>12.3674911660777</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B73V4_ctg10</td>\n",
       "      <td>292384</td>\n",
       "      <td>G</td>\n",
       "      <td>A(2)T(0)C(0)G(906)</td>\n",
       "      <td>51</td>\n",
       "      <td>908</td>\n",
       "      <td>0.220264317180617</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>99.7797356828194</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>B73V4_ctg10</td>\n",
       "      <td>292386</td>\n",
       "      <td>G</td>\n",
       "      <td>A(107)T(0)C(0)G(794)</td>\n",
       "      <td>58</td>\n",
       "      <td>901</td>\n",
       "      <td>11.8756936736959</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>88.1243063263041</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B73V4_ctg10</td>\n",
       "      <td>292402</td>\n",
       "      <td>A</td>\n",
       "      <td>A(912)T(1)C(0)G(0)</td>\n",
       "      <td>46</td>\n",
       "      <td>913</td>\n",
       "      <td>99.8904709748083</td>\n",
       "      <td>0.109529025191676</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>B73V4_ctg129</td>\n",
       "      <td>45198</td>\n",
       "      <td>T</td>\n",
       "      <td>A(0)T(657)C(1)G(0)</td>\n",
       "      <td>301</td>\n",
       "      <td>658</td>\n",
       "      <td>.</td>\n",
       "      <td>99.8480243161094</td>\n",
       "      <td>0.151975683890578</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>B73V4_ctg129</td>\n",
       "      <td>45233</td>\n",
       "      <td>A</td>\n",
       "      <td>A(279)T(0)C(139)G(0)</td>\n",
       "      <td>541</td>\n",
       "      <td>418</td>\n",
       "      <td>66.7464114832536</td>\n",
       "      <td>.</td>\n",
       "      <td>33.2535885167464</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>B73V4_ctg129</td>\n",
       "      <td>45240</td>\n",
       "      <td>A</td>\n",
       "      <td>A(274)T(0)C(104)G(0)</td>\n",
       "      <td>581</td>\n",
       "      <td>378</td>\n",
       "      <td>72.4867724867725</td>\n",
       "      <td>.</td>\n",
       "      <td>27.5132275132275</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>B73V4_ctg129</td>\n",
       "      <td>45257</td>\n",
       "      <td>A</td>\n",
       "      <td>A(286)T(0)C(0)G(200)</td>\n",
       "      <td>473</td>\n",
       "      <td>486</td>\n",
       "      <td>58.8477366255144</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>41.1522633744856</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>B73V4_ctg129</td>\n",
       "      <td>45275</td>\n",
       "      <td>C</td>\n",
       "      <td>A(0)T(35)C(616)G(0)</td>\n",
       "      <td>308</td>\n",
       "      <td>651</td>\n",
       "      <td>.</td>\n",
       "      <td>5.37634408602151</td>\n",
       "      <td>94.6236559139785</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows Ã— 952 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome  position reference_allele         Allele Counts  \\\n",
       "0    B73V4_ctg102     36353                A    A(0)T(0)C(0)G(264)   \n",
       "1     B73V4_ctg10    292327                T  A(0)T(744)C(0)G(105)   \n",
       "2     B73V4_ctg10    292384                G    A(2)T(0)C(0)G(906)   \n",
       "3     B73V4_ctg10    292386                G  A(107)T(0)C(0)G(794)   \n",
       "4     B73V4_ctg10    292402                A    A(912)T(1)C(0)G(0)   \n",
       "..            ...       ...              ...                   ...   \n",
       "995  B73V4_ctg129     45198                T    A(0)T(657)C(1)G(0)   \n",
       "996  B73V4_ctg129     45233                A  A(279)T(0)C(139)G(0)   \n",
       "997  B73V4_ctg129     45240                A  A(274)T(0)C(104)G(0)   \n",
       "998  B73V4_ctg129     45257                A  A(286)T(0)C(0)G(200)   \n",
       "999  B73V4_ctg129     45275                C   A(0)T(35)C(616)G(0)   \n",
       "\n",
       "     Missing Data  Total Calls          Percent A          Percent T  \\\n",
       "0             695          264                  .                  .   \n",
       "1             110          849                  .   87.6325088339223   \n",
       "2              51          908  0.220264317180617                  .   \n",
       "3              58          901   11.8756936736959                  .   \n",
       "4              46          913   99.8904709748083  0.109529025191676   \n",
       "..            ...          ...                ...                ...   \n",
       "995           301          658                  .   99.8480243161094   \n",
       "996           541          418   66.7464114832536                  .   \n",
       "997           581          378   72.4867724867725                  .   \n",
       "998           473          486   58.8477366255144                  .   \n",
       "999           308          651                  .   5.37634408602151   \n",
       "\n",
       "             Percent C         Percent G  ... NC328 NC326 PHV53 DKIBC2 A641  \\\n",
       "0                    .               100  ...     .     .     G      G    G   \n",
       "1                    .  12.3674911660777  ...     T     T     .      T    T   \n",
       "2                    .  99.7797356828194  ...     G     G     G      G    G   \n",
       "3                    .  88.1243063263041  ...     G     G     G      G    G   \n",
       "4                    .                 .  ...     A     A     A      A    A   \n",
       "..                 ...               ...  ...   ...   ...   ...    ...  ...   \n",
       "995  0.151975683890578                 .  ...     .     .     .      T    .   \n",
       "996   33.2535885167464                 .  ...     .     .     .      A    .   \n",
       "997   27.5132275132275                 .  ...     .     .     .      A    .   \n",
       "998                  .  41.1522633744856  ...     .     .     .      A    .   \n",
       "999   94.6236559139785                 .  ...     .     .     .      C    .   \n",
       "\n",
       "    WIL900 Va22 E8501 PHP85 Oh43  \n",
       "0        .    .     G     G    G  \n",
       "1        .    G     T     T    G  \n",
       "2        G    G     G     G    G  \n",
       "3        G    A     G     G    A  \n",
       "4        A    A     A     A    A  \n",
       "..     ...  ...   ...   ...  ...  \n",
       "995      .    .     T     T    T  \n",
       "996      .    .     A     .    C  \n",
       "997      .    .     A     .    C  \n",
       "998      .    .     A     A    G  \n",
       "999      .    .     C     C    C  \n",
       "\n",
       "[1000 rows x 952 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_matrix_raw"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The imputed matrix has 979873 rows Ã— 953 columns. It's structure is different than the one from the raw matrix. Nonetheless, the chromosome and position information is still available.\n",
    "\n",
    "Keep in mind that some nucleotides are reported as `N`.\n",
    "\n",
    "> That is normal. When the sequencing call is ambiguous (like maybe they read multiple possible conflicting nucleotides at that position even though these are inbreds, or the read quality was really low, etc) they use `N` to denote this. Full imputation would convert these to a specific call. So, even though this file has been imputed, it is only a partial imputation to fill in completely missing scores but not ambiguous ones."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs</th>\n",
       "      <th>alleles</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>assembly</th>\n",
       "      <th>center</th>\n",
       "      <th>protLSID</th>\n",
       "      <th>assayLSID</th>\n",
       "      <th>panel</th>\n",
       "      <th>...</th>\n",
       "      <th>YANG</th>\n",
       "      <th>YE_4</th>\n",
       "      <th>YE-CHI-HUNG</th>\n",
       "      <th>YELLOW_3-4</th>\n",
       "      <th>YING-55</th>\n",
       "      <th>Yong_28</th>\n",
       "      <th>Yu796_NS</th>\n",
       "      <th>ZS01250</th>\n",
       "      <th>ZS1791</th>\n",
       "      <th>ZS635</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1_44306</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>44306</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1_44441</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>44441</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1_44879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>44879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1_45948</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>45948</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1_46521</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>46521</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899779</th>\n",
       "      <td>rs10_150833697</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>150833697</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899780</th>\n",
       "      <td>rs10_150833709</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>150833709</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899781</th>\n",
       "      <td>rs10_150833722</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>150833722</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899782</th>\n",
       "      <td>rs10_150833724</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>150833724</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899783</th>\n",
       "      <td>rs10_150833806</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>150833806</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>899784 rows Ã— 953 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    rs  alleles  chrom        pos  strand  assembly  center  \\\n",
       "0            rs1_44306      NaN      1      44306     NaN       NaN     NaN   \n",
       "1            rs1_44441      NaN      1      44441     NaN       NaN     NaN   \n",
       "2            rs1_44879      NaN      1      44879     NaN       NaN     NaN   \n",
       "3            rs1_45948      NaN      1      45948     NaN       NaN     NaN   \n",
       "4            rs1_46521      NaN      1      46521     NaN       NaN     NaN   \n",
       "...                ...      ...    ...        ...     ...       ...     ...   \n",
       "899779  rs10_150833697      NaN     10  150833697     NaN       NaN     NaN   \n",
       "899780  rs10_150833709      NaN     10  150833709     NaN       NaN     NaN   \n",
       "899781  rs10_150833722      NaN     10  150833722     NaN       NaN     NaN   \n",
       "899782  rs10_150833724      NaN     10  150833724     NaN       NaN     NaN   \n",
       "899783  rs10_150833806      NaN     10  150833806     NaN       NaN     NaN   \n",
       "\n",
       "        protLSID  assayLSID  panel  ...  YANG YE_4 YE-CHI-HUNG YELLOW_3-4  \\\n",
       "0            NaN        NaN    NaN  ...     G    G           G          G   \n",
       "1            NaN        NaN    NaN  ...     C    A           A          A   \n",
       "2            NaN        NaN    NaN  ...     C    C           C          C   \n",
       "3            NaN        NaN    NaN  ...     C    C           C          C   \n",
       "4            NaN        NaN    NaN  ...     T    T           T          T   \n",
       "...          ...        ...    ...  ...   ...  ...         ...        ...   \n",
       "899779       NaN        NaN    NaN  ...     C    C           C          C   \n",
       "899780       NaN        NaN    NaN  ...     T    T           T          T   \n",
       "899781       NaN        NaN    NaN  ...     G    G           G          G   \n",
       "899782       NaN        NaN    NaN  ...     T    T           T          T   \n",
       "899783       NaN        NaN    NaN  ...     G    G           G          G   \n",
       "\n",
       "       YING-55 Yong_28 Yu796_NS ZS01250 ZS1791 ZS635  \n",
       "0            G       G        G       G      G     G  \n",
       "1            A       A        A       A      A     A  \n",
       "2            C       C        C       C      C     C  \n",
       "3            C       C        C       C      C     T  \n",
       "4            T       T        T       T      T     T  \n",
       "...        ...     ...      ...     ...    ...   ...  \n",
       "899779       C       C        C       C      C     C  \n",
       "899780       T       T        T       T      T     T  \n",
       "899781       G       G        G       G      G     G  \n",
       "899782       T       T        T       T      T     T  \n",
       "899783       G       G        G       G      G     G  \n",
       "\n",
       "[899784 rows x 953 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_matrix_imputed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Discard the chromosomes `11` and `12`. From the README (emphasis added):\n",
    "> Imputed SNP calls B73 plus B73-derived novel transcripts. SNPs on contigs __that are not__ part of the 10 chromosomes are coded as being on chromosome 11, and SNPs on RTAs are coded as being on chromosome 12.\n",
    "\twidiv_942g_column_name_converter.txt - Converts genotype names between unimputed (column 'Original') and imputed (column 'NoSpaces')."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "if not clean:\n",
    "    downIndex = np.array([])\n",
    "    for i in range(1,11):\n",
    "        chromo = snp_matrix_imputed[snp_matrix_imputed['chrom'] == i]\n",
    "        downIndex = np.concatenate((downIndex, chromo.index.values))\n",
    "\n",
    "    snp_matrix = snp_matrix_imputed.iloc[downIndex]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not clean:\n",
    "    snp_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now have 899784 rows Ã— 953 columns in our cleaned matrix. Observe the math."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not clean:\n",
    "    print('The cleaned matrix has', snp_matrix_imputed.shape[0] - snp_matrix.shape[0], \n",
    "          'rows less than the original imputed matrix.')\n",
    "    print('Observe that the last entry of the cleaned matrix has index', snp_matrix.index[-1],\n",
    "         'while we have a total of', snp_matrix.shape[0], 'rows.')\n",
    "    print('Thus, for future reference, we only need to read the first', snp_matrix.shape[0], \n",
    "          'rows from the original matrix.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternatively, we can read the first 10 individual SNPs from the source file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not clean:\n",
    "    sample_snps = []\n",
    "    with open(src+file_snp_imp) as f:\n",
    "        for i in range(11):\n",
    "            sample_snps.append(f.readline())\n",
    "\n",
    "    # Column names\n",
    "    print(sample_snps[0])\n",
    "\n",
    "    # First SNP\n",
    "    print(sample_snps[1])\n",
    "\n",
    "    # Tenth SNP\n",
    "    print(sample_snps[10])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mask the 10th chromosome to have a smaller dataset to work with. The `rs` column can be reconstructed from the position column. We can also drop the blank columns. \n",
    "\n",
    "> That's because [the original matrix] is a hmp file (hapmap) and those columns are required to be there even though they are seldom if ever used.\n",
    "\n",
    "Now we only have 61462 rows to work with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "downIndex = np.array([])\n",
    "for i in range(10,11):\n",
    "    chromo = snp_matrix_imputed[snp_matrix_imputed['chrom'] == i]\n",
    "    downIndex = np.concatenate((downIndex, chromo.index.values))\n",
    "\n",
    "downIndex.astype(int)\n",
    "snp_sample = snp_matrix_imputed.iloc[downIndex]\n",
    "snp_sample = snp_sample.drop(columns=['rs','alleles', 'strand', \n",
    "                                      'assembly','center', 'protLSID', \n",
    "                                      'assayLSID','panel', 'QCcode'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>764</th>\n",
       "      <th>779</th>\n",
       "      <th>787</th>\n",
       "      <th>790</th>\n",
       "      <th>793</th>\n",
       "      <th>904</th>\n",
       "      <th>911</th>\n",
       "      <th>912</th>\n",
       "      <th>...</th>\n",
       "      <th>YANG</th>\n",
       "      <th>YE_4</th>\n",
       "      <th>YE-CHI-HUNG</th>\n",
       "      <th>YELLOW_3-4</th>\n",
       "      <th>YING-55</th>\n",
       "      <th>Yong_28</th>\n",
       "      <th>Yu796_NS</th>\n",
       "      <th>ZS01250</th>\n",
       "      <th>ZS1791</th>\n",
       "      <th>ZS635</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>838322</th>\n",
       "      <td>10</td>\n",
       "      <td>130347</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>838323</th>\n",
       "      <td>10</td>\n",
       "      <td>130350</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>838324</th>\n",
       "      <td>10</td>\n",
       "      <td>130488</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>838325</th>\n",
       "      <td>10</td>\n",
       "      <td>130523</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>838326</th>\n",
       "      <td>10</td>\n",
       "      <td>130656</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899779</th>\n",
       "      <td>10</td>\n",
       "      <td>150833697</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899780</th>\n",
       "      <td>10</td>\n",
       "      <td>150833709</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899781</th>\n",
       "      <td>10</td>\n",
       "      <td>150833722</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899782</th>\n",
       "      <td>10</td>\n",
       "      <td>150833724</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899783</th>\n",
       "      <td>10</td>\n",
       "      <td>150833806</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61462 rows Ã— 944 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom        pos 764 779 787 790 793 904 911 912  ... YANG YE_4  \\\n",
       "838322     10     130347   C   C   C   C   C   C   C   C  ...    C    C   \n",
       "838323     10     130350   C   C   T   C   C   T   T   T  ...    T    T   \n",
       "838324     10     130488   G   G   C   G   G   C   C   C  ...    C    C   \n",
       "838325     10     130523   T   C   C   T   T   C   C   C  ...    C    C   \n",
       "838326     10     130656   T   T   C   T   T   C   C   C  ...    C    C   \n",
       "...       ...        ...  ..  ..  ..  ..  ..  ..  ..  ..  ...  ...  ...   \n",
       "899779     10  150833697   C   C   C   C   C   C   C   C  ...    C    C   \n",
       "899780     10  150833709   T   T   T   T   T   T   T   T  ...    T    T   \n",
       "899781     10  150833722   G   G   G   G   G   G   G   G  ...    G    G   \n",
       "899782     10  150833724   T   T   T   T   T   T   T   T  ...    T    T   \n",
       "899783     10  150833806   G   G   G   G   G   G   G   G  ...    G    G   \n",
       "\n",
       "       YE-CHI-HUNG YELLOW_3-4 YING-55 Yong_28 Yu796_NS ZS01250 ZS1791 ZS635  \n",
       "838322           C          C       C       C        C       C      C     C  \n",
       "838323           T          T       T       T        T       T      C     T  \n",
       "838324           C          C       C       C        C       C      G     C  \n",
       "838325           C          C       C       C        C       C      T     C  \n",
       "838326           C          C       C       C        C       C      T     C  \n",
       "...            ...        ...     ...     ...      ...     ...    ...   ...  \n",
       "899779           C          C       C       C        C       C      C     C  \n",
       "899780           T          T       T       T        T       T      T     T  \n",
       "899781           G          G       G       G        G       G      G     G  \n",
       "899782           T          T       T       T        T       T      T     T  \n",
       "899783           G          G       G       G        G       G      G     G  \n",
       "\n",
       "[61462 rows x 944 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_sample"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the reduced matrix. Not as small as I expected but not terribly large either. Alternatively, we can further reduce the matrix to consider only the first 1000 SNPs in the 10th chromosome. This downsampled data can be comfortably loaded in any spreadsheet editor (Gnumeric, LibreCalc, MS Excel, etc).\n",
    "\n",
    "If you couldn't load the massive matrix at the very beginning, you can __look for `snp_imputed_chr10_sample.csv` in the `sample_data` directory in git!!!__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_sample.to_csv(dst+'snp_imputed_chr10.csv', index=False)\n",
    "snp_sample.iloc[27000:32000].to_csv(dst+'snp_imputed_chr10_sample.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's explore the expression data!!\n",
    "\n",
    "We have 44259 rows Ã— 947 columns in total.\n",
    "- Rows: FPKM for the same gene across all 942 cultivars\n",
    "- Columns: 5 columns of metadata + 942 cultivars"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ejam/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>position_left</th>\n",
       "      <th>position_right</th>\n",
       "      <th>LH128</th>\n",
       "      <th>DKMBZA</th>\n",
       "      <th>CQ806</th>\n",
       "      <th>DKF274</th>\n",
       "      <th>Ill.Hy</th>\n",
       "      <th>...</th>\n",
       "      <th>NC328</th>\n",
       "      <th>NC326</th>\n",
       "      <th>PHV53</th>\n",
       "      <th>DKIBC2</th>\n",
       "      <th>A641</th>\n",
       "      <th>WIL900</th>\n",
       "      <th>Va22</th>\n",
       "      <th>E8501</th>\n",
       "      <th>PHP85</th>\n",
       "      <th>Oh43</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d012719</td>\n",
       "      <td>8</td>\n",
       "      <td>gene</td>\n",
       "      <td>179164454</td>\n",
       "      <td>179168169</td>\n",
       "      <td>5.825000</td>\n",
       "      <td>4.845710</td>\n",
       "      <td>6.323400</td>\n",
       "      <td>6.999010</td>\n",
       "      <td>6.661100</td>\n",
       "      <td>...</td>\n",
       "      <td>15.154700</td>\n",
       "      <td>12.870700</td>\n",
       "      <td>8.81507</td>\n",
       "      <td>13.77870</td>\n",
       "      <td>8.31790</td>\n",
       "      <td>12.78320</td>\n",
       "      <td>4.94950</td>\n",
       "      <td>9.39900</td>\n",
       "      <td>10.41090</td>\n",
       "      <td>15.840500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Zm00001d024742</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>85863323</td>\n",
       "      <td>85863746</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Zm00001d007137</td>\n",
       "      <td>2</td>\n",
       "      <td>tRNA_gene</td>\n",
       "      <td>223153553</td>\n",
       "      <td>223153627</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Zm00001d025653</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>125107973</td>\n",
       "      <td>125113328</td>\n",
       "      <td>9.350780</td>\n",
       "      <td>9.954800</td>\n",
       "      <td>7.735490</td>\n",
       "      <td>10.360900</td>\n",
       "      <td>9.072890</td>\n",
       "      <td>...</td>\n",
       "      <td>4.685600</td>\n",
       "      <td>3.870230</td>\n",
       "      <td>5.27849</td>\n",
       "      <td>5.17829</td>\n",
       "      <td>3.80291</td>\n",
       "      <td>6.44675</td>\n",
       "      <td>5.61858</td>\n",
       "      <td>3.98055</td>\n",
       "      <td>4.17545</td>\n",
       "      <td>4.500850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Zm00001d036003</td>\n",
       "      <td>6</td>\n",
       "      <td>gene</td>\n",
       "      <td>65322472</td>\n",
       "      <td>65323171</td>\n",
       "      <td>8.955580</td>\n",
       "      <td>0.333921</td>\n",
       "      <td>5.445820</td>\n",
       "      <td>0.332438</td>\n",
       "      <td>1.232330</td>\n",
       "      <td>...</td>\n",
       "      <td>0.232934</td>\n",
       "      <td>0.721774</td>\n",
       "      <td>2.95424</td>\n",
       "      <td>8.44716</td>\n",
       "      <td>12.33680</td>\n",
       "      <td>1.19629</td>\n",
       "      <td>1.71452</td>\n",
       "      <td>7.01768</td>\n",
       "      <td>11.75760</td>\n",
       "      <td>0.793672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44254</th>\n",
       "      <td>Zm00001d028653</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>41833737</td>\n",
       "      <td>41838204</td>\n",
       "      <td>368.777000</td>\n",
       "      <td>301.349000</td>\n",
       "      <td>412.319000</td>\n",
       "      <td>310.820000</td>\n",
       "      <td>278.737000</td>\n",
       "      <td>...</td>\n",
       "      <td>172.293000</td>\n",
       "      <td>186.552000</td>\n",
       "      <td>147.63300</td>\n",
       "      <td>165.60800</td>\n",
       "      <td>147.58300</td>\n",
       "      <td>244.82100</td>\n",
       "      <td>171.15400</td>\n",
       "      <td>219.07400</td>\n",
       "      <td>147.01700</td>\n",
       "      <td>165.612000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44255</th>\n",
       "      <td>Zm00001d009511</td>\n",
       "      <td>8</td>\n",
       "      <td>gene</td>\n",
       "      <td>68418436</td>\n",
       "      <td>68421217</td>\n",
       "      <td>6.003590</td>\n",
       "      <td>5.297160</td>\n",
       "      <td>10.421700</td>\n",
       "      <td>6.304130</td>\n",
       "      <td>4.856300</td>\n",
       "      <td>...</td>\n",
       "      <td>4.120700</td>\n",
       "      <td>4.670860</td>\n",
       "      <td>3.47752</td>\n",
       "      <td>4.92093</td>\n",
       "      <td>6.26904</td>\n",
       "      <td>4.95873</td>\n",
       "      <td>7.25625</td>\n",
       "      <td>5.66864</td>\n",
       "      <td>4.28185</td>\n",
       "      <td>5.765310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44256</th>\n",
       "      <td>Zm00001d032828</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>239455289</td>\n",
       "      <td>239464371</td>\n",
       "      <td>151.479000</td>\n",
       "      <td>46.479600</td>\n",
       "      <td>65.075600</td>\n",
       "      <td>70.169900</td>\n",
       "      <td>59.555600</td>\n",
       "      <td>...</td>\n",
       "      <td>100.249000</td>\n",
       "      <td>102.226000</td>\n",
       "      <td>100.29500</td>\n",
       "      <td>64.90400</td>\n",
       "      <td>91.17440</td>\n",
       "      <td>101.80300</td>\n",
       "      <td>108.58200</td>\n",
       "      <td>63.72140</td>\n",
       "      <td>95.96760</td>\n",
       "      <td>56.404900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44257</th>\n",
       "      <td>Zm00001d002031</td>\n",
       "      <td>2</td>\n",
       "      <td>gene</td>\n",
       "      <td>4673452</td>\n",
       "      <td>4674247</td>\n",
       "      <td>0.237774</td>\n",
       "      <td>0.386886</td>\n",
       "      <td>0.179111</td>\n",
       "      <td>0.415566</td>\n",
       "      <td>0.192793</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1.51552</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>2.555730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44258</th>\n",
       "      <td>Zm00001d011196</td>\n",
       "      <td>8</td>\n",
       "      <td>tRNA_gene</td>\n",
       "      <td>142721025</td>\n",
       "      <td>142721097</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>44259 rows Ã— 947 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene chromosome feature_type  position_left  position_right  \\\n",
       "0      Zm00001d012719          8         gene      179164454       179168169   \n",
       "1      Zm00001d024742         10         gene       85863323        85863746   \n",
       "2      Zm00001d007137          2    tRNA_gene      223153553       223153627   \n",
       "3      Zm00001d025653         10         gene      125107973       125113328   \n",
       "4      Zm00001d036003          6         gene       65322472        65323171   \n",
       "...               ...        ...          ...            ...             ...   \n",
       "44254  Zm00001d028653          1         gene       41833737        41838204   \n",
       "44255  Zm00001d009511          8         gene       68418436        68421217   \n",
       "44256  Zm00001d032828          1         gene      239455289       239464371   \n",
       "44257  Zm00001d002031          2         gene        4673452         4674247   \n",
       "44258  Zm00001d011196          8    tRNA_gene      142721025       142721097   \n",
       "\n",
       "            LH128      DKMBZA       CQ806      DKF274      Ill.Hy  ...  \\\n",
       "0        5.825000    4.845710    6.323400    6.999010    6.661100  ...   \n",
       "1        0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "2        0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "3        9.350780    9.954800    7.735490   10.360900    9.072890  ...   \n",
       "4        8.955580    0.333921    5.445820    0.332438    1.232330  ...   \n",
       "...           ...         ...         ...         ...         ...  ...   \n",
       "44254  368.777000  301.349000  412.319000  310.820000  278.737000  ...   \n",
       "44255    6.003590    5.297160   10.421700    6.304130    4.856300  ...   \n",
       "44256  151.479000   46.479600   65.075600   70.169900   59.555600  ...   \n",
       "44257    0.237774    0.386886    0.179111    0.415566    0.192793  ...   \n",
       "44258    0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "\n",
       "            NC328       NC326      PHV53     DKIBC2       A641     WIL900  \\\n",
       "0       15.154700   12.870700    8.81507   13.77870    8.31790   12.78320   \n",
       "1        0.000000    0.000000    0.00000    0.00000    0.00000    0.00000   \n",
       "2        0.000000    0.000000    0.00000    0.00000    0.00000    0.00000   \n",
       "3        4.685600    3.870230    5.27849    5.17829    3.80291    6.44675   \n",
       "4        0.232934    0.721774    2.95424    8.44716   12.33680    1.19629   \n",
       "...           ...         ...        ...        ...        ...        ...   \n",
       "44254  172.293000  186.552000  147.63300  165.60800  147.58300  244.82100   \n",
       "44255    4.120700    4.670860    3.47752    4.92093    6.26904    4.95873   \n",
       "44256  100.249000  102.226000  100.29500   64.90400   91.17440  101.80300   \n",
       "44257    0.000000    0.000000    0.00000    0.00000    0.00000    1.51552   \n",
       "44258    0.000000    0.000000    0.00000    0.00000    0.00000    0.00000   \n",
       "\n",
       "            Va22      E8501      PHP85        Oh43  \n",
       "0        4.94950    9.39900   10.41090   15.840500  \n",
       "1        0.00000    0.00000    0.00000    0.000000  \n",
       "2        0.00000    0.00000    0.00000    0.000000  \n",
       "3        5.61858    3.98055    4.17545    4.500850  \n",
       "4        1.71452    7.01768   11.75760    0.793672  \n",
       "...          ...        ...        ...         ...  \n",
       "44254  171.15400  219.07400  147.01700  165.612000  \n",
       "44255    7.25625    5.66864    4.28185    5.765310  \n",
       "44256  108.58200   63.72140   95.96760   56.404900  \n",
       "44257    0.00000    0.00000    0.00000    2.555730  \n",
       "44258    0.00000    0.00000    0.00000    0.000000  \n",
       "\n",
       "[44259 rows x 947 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_w_features_raw = pd.read_table(src+file_gen)\n",
    "genes_w_features_raw"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Keep only information of the genes in chromosomes. Ignore the contigs. \n",
    "\n",
    "Turns out the contigs are mingled with the chromosomes and we can't define a nice cutoff while reading. We must read the whole file and then filter, as below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "downIndex = np.array([])\n",
    "for i in range(1,11):\n",
    "    chromo = genes_w_features_raw[genes_w_features_raw['chromosome'] == '{}'.format(i)]\n",
    "    downIndex = np.concatenate((downIndex, chromo.index.values))\n",
    "\n",
    "downIndex.astype(int)\n",
    "gene_w_features = genes_w_features_raw.iloc[downIndex].astype({'chromosome': int})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>position_left</th>\n",
       "      <th>position_right</th>\n",
       "      <th>LH128</th>\n",
       "      <th>DKMBZA</th>\n",
       "      <th>CQ806</th>\n",
       "      <th>DKF274</th>\n",
       "      <th>Ill.Hy</th>\n",
       "      <th>...</th>\n",
       "      <th>NC328</th>\n",
       "      <th>NC326</th>\n",
       "      <th>PHV53</th>\n",
       "      <th>DKIBC2</th>\n",
       "      <th>A641</th>\n",
       "      <th>WIL900</th>\n",
       "      <th>Va22</th>\n",
       "      <th>E8501</th>\n",
       "      <th>PHP85</th>\n",
       "      <th>Oh43</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Zm00001d033979</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>279975978</td>\n",
       "      <td>279978430</td>\n",
       "      <td>76.133600</td>\n",
       "      <td>65.408400</td>\n",
       "      <td>38.507300</td>\n",
       "      <td>52.069200</td>\n",
       "      <td>75.549300</td>\n",
       "      <td>...</td>\n",
       "      <td>42.306200</td>\n",
       "      <td>41.304400</td>\n",
       "      <td>36.242400</td>\n",
       "      <td>33.992500</td>\n",
       "      <td>30.771400</td>\n",
       "      <td>23.95840</td>\n",
       "      <td>24.665400</td>\n",
       "      <td>20.311500</td>\n",
       "      <td>17.770000</td>\n",
       "      <td>38.619700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Zm00001d031073</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>175980668</td>\n",
       "      <td>175983357</td>\n",
       "      <td>1.381800</td>\n",
       "      <td>0.621732</td>\n",
       "      <td>0.852335</td>\n",
       "      <td>1.681980</td>\n",
       "      <td>0.958799</td>\n",
       "      <td>...</td>\n",
       "      <td>1.341690</td>\n",
       "      <td>1.241070</td>\n",
       "      <td>0.877111</td>\n",
       "      <td>1.106620</td>\n",
       "      <td>1.758750</td>\n",
       "      <td>2.21258</td>\n",
       "      <td>1.086100</td>\n",
       "      <td>1.034650</td>\n",
       "      <td>2.463450</td>\n",
       "      <td>0.652187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Zm00001d034961</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>306881093</td>\n",
       "      <td>306895296</td>\n",
       "      <td>0.026991</td>\n",
       "      <td>0.005325</td>\n",
       "      <td>0.016174</td>\n",
       "      <td>0.011593</td>\n",
       "      <td>0.018716</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009031</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015687</td>\n",
       "      <td>0.030662</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.019362</td>\n",
       "      <td>0.018610</td>\n",
       "      <td>0.034208</td>\n",
       "      <td>0.026505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Zm00001d022701</td>\n",
       "      <td>1</td>\n",
       "      <td>lincRNA_gene</td>\n",
       "      <td>27303532</td>\n",
       "      <td>27304500</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.040027</td>\n",
       "      <td>0.018941</td>\n",
       "      <td>0.088167</td>\n",
       "      <td>0.060270</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.181800</td>\n",
       "      <td>0.110069</td>\n",
       "      <td>0.14295</td>\n",
       "      <td>0.111420</td>\n",
       "      <td>0.110546</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.074827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Zm00001d022951</td>\n",
       "      <td>1</td>\n",
       "      <td>lincRNA_gene</td>\n",
       "      <td>247966116</td>\n",
       "      <td>247966918</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43951</th>\n",
       "      <td>Zm00001d023459</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>6289475</td>\n",
       "      <td>6298739</td>\n",
       "      <td>7.743020</td>\n",
       "      <td>12.641800</td>\n",
       "      <td>10.975500</td>\n",
       "      <td>12.337900</td>\n",
       "      <td>15.497200</td>\n",
       "      <td>...</td>\n",
       "      <td>7.251110</td>\n",
       "      <td>13.711400</td>\n",
       "      <td>15.806400</td>\n",
       "      <td>16.518300</td>\n",
       "      <td>6.885620</td>\n",
       "      <td>22.86330</td>\n",
       "      <td>8.189910</td>\n",
       "      <td>6.772000</td>\n",
       "      <td>7.192290</td>\n",
       "      <td>10.332200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43976</th>\n",
       "      <td>Zm00001d026184</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>140529821</td>\n",
       "      <td>140530718</td>\n",
       "      <td>0.437610</td>\n",
       "      <td>0.361114</td>\n",
       "      <td>0.366553</td>\n",
       "      <td>0.512719</td>\n",
       "      <td>0.149871</td>\n",
       "      <td>...</td>\n",
       "      <td>0.742883</td>\n",
       "      <td>0.587186</td>\n",
       "      <td>2.192500</td>\n",
       "      <td>0.110062</td>\n",
       "      <td>1.083900</td>\n",
       "      <td>1.57213</td>\n",
       "      <td>0.272518</td>\n",
       "      <td>0.154927</td>\n",
       "      <td>0.975337</td>\n",
       "      <td>0.605777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43984</th>\n",
       "      <td>Zm00001d024000</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>35675291</td>\n",
       "      <td>35679293</td>\n",
       "      <td>38.797700</td>\n",
       "      <td>28.401200</td>\n",
       "      <td>3.048510</td>\n",
       "      <td>4.423800</td>\n",
       "      <td>8.698070</td>\n",
       "      <td>...</td>\n",
       "      <td>0.093581</td>\n",
       "      <td>2.816500</td>\n",
       "      <td>5.026650</td>\n",
       "      <td>3.186000</td>\n",
       "      <td>0.373411</td>\n",
       "      <td>18.87170</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.644770</td>\n",
       "      <td>33.329400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44023</th>\n",
       "      <td>Zm00001d025251</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>111149588</td>\n",
       "      <td>111152196</td>\n",
       "      <td>0.081500</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.068706</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.046316</td>\n",
       "      <td>0.062825</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.252784</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.033119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44030</th>\n",
       "      <td>Zm00001d025948</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>134519363</td>\n",
       "      <td>134523011</td>\n",
       "      <td>53.629600</td>\n",
       "      <td>45.164900</td>\n",
       "      <td>80.306500</td>\n",
       "      <td>48.090300</td>\n",
       "      <td>45.858400</td>\n",
       "      <td>...</td>\n",
       "      <td>31.725100</td>\n",
       "      <td>29.258700</td>\n",
       "      <td>31.312400</td>\n",
       "      <td>35.837800</td>\n",
       "      <td>38.467700</td>\n",
       "      <td>23.32740</td>\n",
       "      <td>33.359300</td>\n",
       "      <td>30.807100</td>\n",
       "      <td>42.896200</td>\n",
       "      <td>38.359100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43675 rows Ã— 947 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene  chromosome  feature_type  position_left  \\\n",
       "7      Zm00001d033979           1          gene      279975978   \n",
       "22     Zm00001d031073           1          gene      175980668   \n",
       "23     Zm00001d034961           1          gene      306881093   \n",
       "35     Zm00001d022701           1  lincRNA_gene       27303532   \n",
       "42     Zm00001d022951           1  lincRNA_gene      247966116   \n",
       "...               ...         ...           ...            ...   \n",
       "43951  Zm00001d023459          10          gene        6289475   \n",
       "43976  Zm00001d026184          10          gene      140529821   \n",
       "43984  Zm00001d024000          10          gene       35675291   \n",
       "44023  Zm00001d025251          10          gene      111149588   \n",
       "44030  Zm00001d025948          10          gene      134519363   \n",
       "\n",
       "       position_right      LH128     DKMBZA      CQ806     DKF274     Ill.Hy  \\\n",
       "7           279978430  76.133600  65.408400  38.507300  52.069200  75.549300   \n",
       "22          175983357   1.381800   0.621732   0.852335   1.681980   0.958799   \n",
       "23          306895296   0.026991   0.005325   0.016174   0.011593   0.018716   \n",
       "35           27304500   0.000000   0.040027   0.018941   0.088167   0.060270   \n",
       "42          247966918   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "...               ...        ...        ...        ...        ...        ...   \n",
       "43951         6298739   7.743020  12.641800  10.975500  12.337900  15.497200   \n",
       "43976       140530718   0.437610   0.361114   0.366553   0.512719   0.149871   \n",
       "43984        35679293  38.797700  28.401200   3.048510   4.423800   8.698070   \n",
       "44023       111152196   0.081500   0.000000   0.068706   0.000000   0.000000   \n",
       "44030       134523011  53.629600  45.164900  80.306500  48.090300  45.858400   \n",
       "\n",
       "       ...      NC328      NC326      PHV53     DKIBC2       A641    WIL900  \\\n",
       "7      ...  42.306200  41.304400  36.242400  33.992500  30.771400  23.95840   \n",
       "22     ...   1.341690   1.241070   0.877111   1.106620   1.758750   2.21258   \n",
       "23     ...   0.009031   0.000000   0.015687   0.030662   0.000000   0.00000   \n",
       "35     ...   0.000000   0.000000   0.000000   0.181800   0.110069   0.14295   \n",
       "42     ...   0.000000   0.000000   0.000000   0.000000   0.000000   0.00000   \n",
       "...    ...        ...        ...        ...        ...        ...       ...   \n",
       "43951  ...   7.251110  13.711400  15.806400  16.518300   6.885620  22.86330   \n",
       "43976  ...   0.742883   0.587186   2.192500   0.110062   1.083900   1.57213   \n",
       "43984  ...   0.093581   2.816500   5.026650   3.186000   0.373411  18.87170   \n",
       "44023  ...   0.000000   0.000000   0.000000   0.046316   0.062825   0.00000   \n",
       "44030  ...  31.725100  29.258700  31.312400  35.837800  38.467700  23.32740   \n",
       "\n",
       "            Va22      E8501      PHP85       Oh43  \n",
       "7      24.665400  20.311500  17.770000  38.619700  \n",
       "22      1.086100   1.034650   2.463450   0.652187  \n",
       "23      0.019362   0.018610   0.034208   0.026505  \n",
       "35      0.111420   0.110546   0.000000   0.074827  \n",
       "42      0.000000   0.000000   0.000000   0.000000  \n",
       "...          ...        ...        ...        ...  \n",
       "43951   8.189910   6.772000   7.192290  10.332200  \n",
       "43976   0.272518   0.154927   0.975337   0.605777  \n",
       "43984   0.000000   0.000000   9.644770  33.329400  \n",
       "44023   0.000000   0.252784   0.000000   0.033119  \n",
       "44030  33.359300  30.807100  42.896200  38.359100  \n",
       "\n",
       "[43675 rows x 947 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_w_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We discarded 584 rows.\n"
     ]
    }
   ],
   "source": [
    "print('We discarded', genes_w_features_raw.shape[0] - gene_w_features.shape[0],'rows.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Downsample the matrix. Consider only the genes in the 10th chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>position_left</th>\n",
       "      <th>position_right</th>\n",
       "      <th>LH128</th>\n",
       "      <th>DKMBZA</th>\n",
       "      <th>CQ806</th>\n",
       "      <th>DKF274</th>\n",
       "      <th>Ill.Hy</th>\n",
       "      <th>...</th>\n",
       "      <th>NC328</th>\n",
       "      <th>NC326</th>\n",
       "      <th>PHV53</th>\n",
       "      <th>DKIBC2</th>\n",
       "      <th>A641</th>\n",
       "      <th>WIL900</th>\n",
       "      <th>Va22</th>\n",
       "      <th>E8501</th>\n",
       "      <th>PHP85</th>\n",
       "      <th>Oh43</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Zm00001d024742</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>85863323</td>\n",
       "      <td>85863746</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Zm00001d025653</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>125107973</td>\n",
       "      <td>125113328</td>\n",
       "      <td>9.350780</td>\n",
       "      <td>9.954800</td>\n",
       "      <td>7.735490</td>\n",
       "      <td>10.360900</td>\n",
       "      <td>9.072890</td>\n",
       "      <td>...</td>\n",
       "      <td>4.685600</td>\n",
       "      <td>3.870230</td>\n",
       "      <td>5.278490</td>\n",
       "      <td>5.178290</td>\n",
       "      <td>3.802910</td>\n",
       "      <td>6.446750</td>\n",
       "      <td>5.618580</td>\n",
       "      <td>3.980550</td>\n",
       "      <td>4.175450</td>\n",
       "      <td>4.500850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Zm00001d024903</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>93908316</td>\n",
       "      <td>93911235</td>\n",
       "      <td>0.954312</td>\n",
       "      <td>0.335773</td>\n",
       "      <td>0.689938</td>\n",
       "      <td>0.106595</td>\n",
       "      <td>0.200660</td>\n",
       "      <td>...</td>\n",
       "      <td>1.809510</td>\n",
       "      <td>1.671710</td>\n",
       "      <td>2.940510</td>\n",
       "      <td>1.523060</td>\n",
       "      <td>7.885560</td>\n",
       "      <td>4.522500</td>\n",
       "      <td>1.478990</td>\n",
       "      <td>9.566990</td>\n",
       "      <td>6.826610</td>\n",
       "      <td>0.509826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Zm00001d023618</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>12298031</td>\n",
       "      <td>12300071</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Zm00001d024825</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>89733155</td>\n",
       "      <td>89739279</td>\n",
       "      <td>0.857062</td>\n",
       "      <td>0.689824</td>\n",
       "      <td>0.613990</td>\n",
       "      <td>1.146330</td>\n",
       "      <td>1.137050</td>\n",
       "      <td>...</td>\n",
       "      <td>0.617702</td>\n",
       "      <td>0.317978</td>\n",
       "      <td>0.354413</td>\n",
       "      <td>0.412837</td>\n",
       "      <td>0.495563</td>\n",
       "      <td>0.312885</td>\n",
       "      <td>0.616760</td>\n",
       "      <td>0.450897</td>\n",
       "      <td>0.610228</td>\n",
       "      <td>0.472077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43951</th>\n",
       "      <td>Zm00001d023459</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>6289475</td>\n",
       "      <td>6298739</td>\n",
       "      <td>7.743020</td>\n",
       "      <td>12.641800</td>\n",
       "      <td>10.975500</td>\n",
       "      <td>12.337900</td>\n",
       "      <td>15.497200</td>\n",
       "      <td>...</td>\n",
       "      <td>7.251110</td>\n",
       "      <td>13.711400</td>\n",
       "      <td>15.806400</td>\n",
       "      <td>16.518300</td>\n",
       "      <td>6.885620</td>\n",
       "      <td>22.863300</td>\n",
       "      <td>8.189910</td>\n",
       "      <td>6.772000</td>\n",
       "      <td>7.192290</td>\n",
       "      <td>10.332200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43976</th>\n",
       "      <td>Zm00001d026184</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>140529821</td>\n",
       "      <td>140530718</td>\n",
       "      <td>0.437610</td>\n",
       "      <td>0.361114</td>\n",
       "      <td>0.366553</td>\n",
       "      <td>0.512719</td>\n",
       "      <td>0.149871</td>\n",
       "      <td>...</td>\n",
       "      <td>0.742883</td>\n",
       "      <td>0.587186</td>\n",
       "      <td>2.192500</td>\n",
       "      <td>0.110062</td>\n",
       "      <td>1.083900</td>\n",
       "      <td>1.572130</td>\n",
       "      <td>0.272518</td>\n",
       "      <td>0.154927</td>\n",
       "      <td>0.975337</td>\n",
       "      <td>0.605777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43984</th>\n",
       "      <td>Zm00001d024000</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>35675291</td>\n",
       "      <td>35679293</td>\n",
       "      <td>38.797700</td>\n",
       "      <td>28.401200</td>\n",
       "      <td>3.048510</td>\n",
       "      <td>4.423800</td>\n",
       "      <td>8.698070</td>\n",
       "      <td>...</td>\n",
       "      <td>0.093581</td>\n",
       "      <td>2.816500</td>\n",
       "      <td>5.026650</td>\n",
       "      <td>3.186000</td>\n",
       "      <td>0.373411</td>\n",
       "      <td>18.871700</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.644770</td>\n",
       "      <td>33.329400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44023</th>\n",
       "      <td>Zm00001d025251</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>111149588</td>\n",
       "      <td>111152196</td>\n",
       "      <td>0.081500</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.068706</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.046316</td>\n",
       "      <td>0.062825</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.252784</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.033119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44030</th>\n",
       "      <td>Zm00001d025948</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>134519363</td>\n",
       "      <td>134523011</td>\n",
       "      <td>53.629600</td>\n",
       "      <td>45.164900</td>\n",
       "      <td>80.306500</td>\n",
       "      <td>48.090300</td>\n",
       "      <td>45.858400</td>\n",
       "      <td>...</td>\n",
       "      <td>31.725100</td>\n",
       "      <td>29.258700</td>\n",
       "      <td>31.312400</td>\n",
       "      <td>35.837800</td>\n",
       "      <td>38.467700</td>\n",
       "      <td>23.327400</td>\n",
       "      <td>33.359300</td>\n",
       "      <td>30.807100</td>\n",
       "      <td>42.896200</td>\n",
       "      <td>38.359100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2987 rows Ã— 947 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene  chromosome feature_type  position_left  position_right  \\\n",
       "1      Zm00001d024742          10         gene       85863323        85863746   \n",
       "3      Zm00001d025653          10         gene      125107973       125113328   \n",
       "6      Zm00001d024903          10         gene       93908316        93911235   \n",
       "11     Zm00001d023618          10         gene       12298031        12300071   \n",
       "14     Zm00001d024825          10         gene       89733155        89739279   \n",
       "...               ...         ...          ...            ...             ...   \n",
       "43951  Zm00001d023459          10         gene        6289475         6298739   \n",
       "43976  Zm00001d026184          10         gene      140529821       140530718   \n",
       "43984  Zm00001d024000          10         gene       35675291        35679293   \n",
       "44023  Zm00001d025251          10         gene      111149588       111152196   \n",
       "44030  Zm00001d025948          10         gene      134519363       134523011   \n",
       "\n",
       "           LH128     DKMBZA      CQ806     DKF274     Ill.Hy  ...      NC328  \\\n",
       "1       0.000000   0.000000   0.000000   0.000000   0.000000  ...   0.000000   \n",
       "3       9.350780   9.954800   7.735490  10.360900   9.072890  ...   4.685600   \n",
       "6       0.954312   0.335773   0.689938   0.106595   0.200660  ...   1.809510   \n",
       "11      0.000000   0.000000   0.000000   0.000000   0.000000  ...   0.000000   \n",
       "14      0.857062   0.689824   0.613990   1.146330   1.137050  ...   0.617702   \n",
       "...          ...        ...        ...        ...        ...  ...        ...   \n",
       "43951   7.743020  12.641800  10.975500  12.337900  15.497200  ...   7.251110   \n",
       "43976   0.437610   0.361114   0.366553   0.512719   0.149871  ...   0.742883   \n",
       "43984  38.797700  28.401200   3.048510   4.423800   8.698070  ...   0.093581   \n",
       "44023   0.081500   0.000000   0.068706   0.000000   0.000000  ...   0.000000   \n",
       "44030  53.629600  45.164900  80.306500  48.090300  45.858400  ...  31.725100   \n",
       "\n",
       "           NC326      PHV53     DKIBC2       A641     WIL900       Va22  \\\n",
       "1       0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "3       3.870230   5.278490   5.178290   3.802910   6.446750   5.618580   \n",
       "6       1.671710   2.940510   1.523060   7.885560   4.522500   1.478990   \n",
       "11      0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "14      0.317978   0.354413   0.412837   0.495563   0.312885   0.616760   \n",
       "...          ...        ...        ...        ...        ...        ...   \n",
       "43951  13.711400  15.806400  16.518300   6.885620  22.863300   8.189910   \n",
       "43976   0.587186   2.192500   0.110062   1.083900   1.572130   0.272518   \n",
       "43984   2.816500   5.026650   3.186000   0.373411  18.871700   0.000000   \n",
       "44023   0.000000   0.000000   0.046316   0.062825   0.000000   0.000000   \n",
       "44030  29.258700  31.312400  35.837800  38.467700  23.327400  33.359300   \n",
       "\n",
       "           E8501      PHP85       Oh43  \n",
       "1       0.000000   0.000000   0.000000  \n",
       "3       3.980550   4.175450   4.500850  \n",
       "6       9.566990   6.826610   0.509826  \n",
       "11      0.000000   0.000000   0.000000  \n",
       "14      0.450897   0.610228   0.472077  \n",
       "...          ...        ...        ...  \n",
       "43951   6.772000   7.192290  10.332200  \n",
       "43976   0.154927   0.975337   0.605777  \n",
       "43984   0.000000   9.644770  33.329400  \n",
       "44023   0.252784   0.000000   0.033119  \n",
       "44030  30.807100  42.896200  38.359100  \n",
       "\n",
       "[2987 rows x 947 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_sample = gene_w_features[gene_w_features['chromosome'] == 10]\n",
    "gene_sample"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you couldn't load the massive matrix at the very beginning, you can __look for `gene_w_features_chr10.csv` in the `sample_data` directory in git!!!__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_sample.to_csv(dst+'gene_w_features_chr10.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>position_left</th>\n",
       "      <th>position_right</th>\n",
       "      <th>LH128</th>\n",
       "      <th>DKMBZA</th>\n",
       "      <th>CQ806</th>\n",
       "      <th>DKF274</th>\n",
       "      <th>Ill.Hy</th>\n",
       "      <th>...</th>\n",
       "      <th>NC328</th>\n",
       "      <th>NC326</th>\n",
       "      <th>PHV53</th>\n",
       "      <th>DKIBC2</th>\n",
       "      <th>A641</th>\n",
       "      <th>WIL900</th>\n",
       "      <th>Va22</th>\n",
       "      <th>E8501</th>\n",
       "      <th>PHP85</th>\n",
       "      <th>Oh43</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>33396</th>\n",
       "      <td>Zm00001d025252</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>111156937</td>\n",
       "      <td>111166083</td>\n",
       "      <td>2.54496</td>\n",
       "      <td>2.37732</td>\n",
       "      <td>2.68808</td>\n",
       "      <td>3.28275</td>\n",
       "      <td>3.22185</td>\n",
       "      <td>...</td>\n",
       "      <td>2.3473</td>\n",
       "      <td>3.18874</td>\n",
       "      <td>2.98677</td>\n",
       "      <td>3.26416</td>\n",
       "      <td>2.44131</td>\n",
       "      <td>4.09</td>\n",
       "      <td>3.53631</td>\n",
       "      <td>3.1623</td>\n",
       "      <td>3.44806</td>\n",
       "      <td>2.47579</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 947 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene  chromosome feature_type  position_left  position_right  \\\n",
       "33396  Zm00001d025252          10         gene      111156937       111166083   \n",
       "\n",
       "         LH128   DKMBZA    CQ806   DKF274   Ill.Hy  ...   NC328    NC326  \\\n",
       "33396  2.54496  2.37732  2.68808  3.28275  3.22185  ...  2.3473  3.18874   \n",
       "\n",
       "         PHV53   DKIBC2     A641  WIL900     Va22   E8501    PHP85     Oh43  \n",
       "33396  2.98677  3.26416  2.44131    4.09  3.53631  3.1623  3.44806  2.47579  \n",
       "\n",
       "[1 rows x 947 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val = 111166032\n",
    "l_match = gene_sample[gene_sample['position_left'] <= val]\n",
    "r_match = l_match[l_match['position_right'] >= val]\n",
    "r_match"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__To Do:__\n",
    "- Match efficiently each SNP to its respective gene based on the position information."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
