{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <p style=\"text-align: right;\"> &#9989; **Group 1 (GoGreen)** </p>\n",
    "#### <p style=\"text-align: right;\"> &#9989; Beth, Zhongjie, McKenna, Erik</p>\n",
    "\n",
    "# Module 1\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How do SNPs relate to gene transcription?\n",
    "\n",
    "We will work with maize. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Jupyter by default converts `tabs` into `spaces` which is good."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, import the usual libraries\n",
    "- `math`: basic math operations\n",
    "- `os`: enable file manipulation with the OS\n",
    "- `sys`: enable interaction with commandline\n",
    "- `glob`: more variable manipulation\n",
    "- `matplotlib.pyplot`: default plotter (I personally like ggplot waaaaay better. E)\n",
    "    - `inline`: so that plots are shown in the notebook\n",
    "- `seaborn`: nicer plots\n",
    "- `numpy`: all number cruching done here\n",
    "- `pandas`: data wrangling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math  \n",
    "import os   \n",
    "import glob \n",
    "import matplotlib.pyplot as plt \n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define some functions. Let's see for how long can we keep tidy commented code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dot_to_int(char_in, substitute='.', value=0.0):\n",
    "    \"\"\"\n",
    "    Substitute a string for a given value if such string is of interest\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    char_in : string\n",
    "         possible char to be substituted\n",
    "    \n",
    "    substitute : string\n",
    "         if the given input matches this object, substitute\n",
    "    \n",
    "    value : int\n",
    "         if applicable, substitute the input with this value\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    value : if input matches substitute, else returns unchanged input\n",
    "    \"\"\"\n",
    "    if(char_in == substitute):\n",
    "        return value\n",
    "    else:\n",
    "        return char_in\n",
    "    \n",
    "def numMissing(row, start=10, missing='.'):\n",
    "    foo = row[start:] == missing\n",
    "    return(foo.sum())\n",
    "\n",
    "\n",
    "def numData(row, data_cols=['Total Calls', 'Missing Data']):\n",
    "    return row[data_cols].sum()  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can print whatever text was written between the `\"\"\"` above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    Substitute a string for a given value if such string is of interest\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    char_in : string\n",
      "         possible char to be substituted\n",
      "    \n",
      "    substitute : string\n",
      "         if the given input matches this object, substitute\n",
      "    \n",
      "    value : int\n",
      "         if applicable, substitute the input with this value\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    value : if input matches substitute, else returns unchanged input\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "print(dot_to_int.__doc__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the paths to the raw data\n",
    "src = '/home/ejam/documents/css893/raw_data/'\n",
    "dst = '/home/ejam/documents/css893/output/'\n",
    "file_snp = 'B73_plus_RTAs_snp_matrix_995785.txt'\n",
    "file_gen = '942_FPKM_B73_genes_w_feature.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "ATCG = ['Percent A', 'Percent T', 'Percent C', 'Percent G']\n",
    "converters = dict((A,dot_to_int) for A in ATCG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# BEWARE : Line below consumes 24Gb of RAM at some point. \n",
    "snp_matrix = pd.read_table(src + file_snp, dtype={'chromosome': str}, converters=converters)\n",
    "\n",
    "#snp_matrix = pd.read_table(src + file_snp, dtype={'chromosome': str}, converters=converters, nrows=1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>Allele Counts</th>\n",
       "      <th>Missing Data</th>\n",
       "      <th>Total Calls</th>\n",
       "      <th>Percent A</th>\n",
       "      <th>Percent T</th>\n",
       "      <th>Percent C</th>\n",
       "      <th>Percent G</th>\n",
       "      <th>...</th>\n",
       "      <th>NC328</th>\n",
       "      <th>NC326</th>\n",
       "      <th>PHV53</th>\n",
       "      <th>DKIBC2</th>\n",
       "      <th>A641</th>\n",
       "      <th>WIL900</th>\n",
       "      <th>Va22</th>\n",
       "      <th>E8501</th>\n",
       "      <th>PHP85</th>\n",
       "      <th>Oh43</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>B73V4_ctg102</td>\n",
       "      <td>36353</td>\n",
       "      <td>A</td>\n",
       "      <td>A(0)T(0)C(0)G(264)</td>\n",
       "      <td>695</td>\n",
       "      <td>264</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B73V4_ctg10</td>\n",
       "      <td>292327</td>\n",
       "      <td>T</td>\n",
       "      <td>A(0)T(744)C(0)G(105)</td>\n",
       "      <td>110</td>\n",
       "      <td>849</td>\n",
       "      <td>0</td>\n",
       "      <td>87.6325088339223</td>\n",
       "      <td>0</td>\n",
       "      <td>12.3674911660777</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B73V4_ctg10</td>\n",
       "      <td>292384</td>\n",
       "      <td>G</td>\n",
       "      <td>A(2)T(0)C(0)G(906)</td>\n",
       "      <td>51</td>\n",
       "      <td>908</td>\n",
       "      <td>0.220264317180617</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>99.7797356828194</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>B73V4_ctg10</td>\n",
       "      <td>292386</td>\n",
       "      <td>G</td>\n",
       "      <td>A(107)T(0)C(0)G(794)</td>\n",
       "      <td>58</td>\n",
       "      <td>901</td>\n",
       "      <td>11.8756936736959</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>88.1243063263041</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B73V4_ctg10</td>\n",
       "      <td>292402</td>\n",
       "      <td>A</td>\n",
       "      <td>A(912)T(1)C(0)G(0)</td>\n",
       "      <td>46</td>\n",
       "      <td>913</td>\n",
       "      <td>99.8904709748083</td>\n",
       "      <td>0.109529025191676</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995780</th>\n",
       "      <td>10</td>\n",
       "      <td>99998392</td>\n",
       "      <td>C</td>\n",
       "      <td>A(0)T(5)C(954)G(0)</td>\n",
       "      <td>0</td>\n",
       "      <td>959</td>\n",
       "      <td>0</td>\n",
       "      <td>0.521376433785193</td>\n",
       "      <td>99.4786235662148</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995781</th>\n",
       "      <td>10</td>\n",
       "      <td>99998427</td>\n",
       "      <td>A</td>\n",
       "      <td>A(855)T(0)C(30)G(0)</td>\n",
       "      <td>74</td>\n",
       "      <td>885</td>\n",
       "      <td>96.6101694915254</td>\n",
       "      <td>0</td>\n",
       "      <td>3.38983050847458</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995782</th>\n",
       "      <td>10</td>\n",
       "      <td>99998428</td>\n",
       "      <td>G</td>\n",
       "      <td>A(7)T(0)C(0)G(857)</td>\n",
       "      <td>95</td>\n",
       "      <td>864</td>\n",
       "      <td>0.810185185185185</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>99.1898148148148</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995783</th>\n",
       "      <td>10</td>\n",
       "      <td>99998430</td>\n",
       "      <td>C</td>\n",
       "      <td>A(0)T(0)C(843)G(5)</td>\n",
       "      <td>111</td>\n",
       "      <td>848</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>99.4103773584906</td>\n",
       "      <td>0.589622641509434</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995784</th>\n",
       "      <td>10</td>\n",
       "      <td>99998441</td>\n",
       "      <td>G</td>\n",
       "      <td>A(437)T(0)C(0)G(136)</td>\n",
       "      <td>386</td>\n",
       "      <td>573</td>\n",
       "      <td>76.2652705061082</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23.7347294938918</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>995785 rows × 952 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chromosome  position reference_allele         Allele Counts  \\\n",
       "0       B73V4_ctg102     36353                A    A(0)T(0)C(0)G(264)   \n",
       "1        B73V4_ctg10    292327                T  A(0)T(744)C(0)G(105)   \n",
       "2        B73V4_ctg10    292384                G    A(2)T(0)C(0)G(906)   \n",
       "3        B73V4_ctg10    292386                G  A(107)T(0)C(0)G(794)   \n",
       "4        B73V4_ctg10    292402                A    A(912)T(1)C(0)G(0)   \n",
       "...              ...       ...              ...                   ...   \n",
       "995780            10  99998392                C    A(0)T(5)C(954)G(0)   \n",
       "995781            10  99998427                A   A(855)T(0)C(30)G(0)   \n",
       "995782            10  99998428                G    A(7)T(0)C(0)G(857)   \n",
       "995783            10  99998430                C    A(0)T(0)C(843)G(5)   \n",
       "995784            10  99998441                G  A(437)T(0)C(0)G(136)   \n",
       "\n",
       "        Missing Data  Total Calls          Percent A          Percent T  \\\n",
       "0                695          264                  0                  0   \n",
       "1                110          849                  0   87.6325088339223   \n",
       "2                 51          908  0.220264317180617                  0   \n",
       "3                 58          901   11.8756936736959                  0   \n",
       "4                 46          913   99.8904709748083  0.109529025191676   \n",
       "...              ...          ...                ...                ...   \n",
       "995780             0          959                  0  0.521376433785193   \n",
       "995781            74          885   96.6101694915254                  0   \n",
       "995782            95          864  0.810185185185185                  0   \n",
       "995783           111          848                  0                  0   \n",
       "995784           386          573   76.2652705061082                  0   \n",
       "\n",
       "               Percent C          Percent G  ... NC328 NC326 PHV53 DKIBC2  \\\n",
       "0                      0                100  ...     .     .     G      G   \n",
       "1                      0   12.3674911660777  ...     T     T     .      T   \n",
       "2                      0   99.7797356828194  ...     G     G     G      G   \n",
       "3                      0   88.1243063263041  ...     G     G     G      G   \n",
       "4                      0                  0  ...     A     A     A      A   \n",
       "...                  ...                ...  ...   ...   ...   ...    ...   \n",
       "995780  99.4786235662148                  0  ...     C     C     C      C   \n",
       "995781  3.38983050847458                  0  ...     A     A     A      A   \n",
       "995782                 0   99.1898148148148  ...     G     G     G      G   \n",
       "995783  99.4103773584906  0.589622641509434  ...     C     C     C      C   \n",
       "995784                 0   23.7347294938918  ...     .     G     G      A   \n",
       "\n",
       "       A641 WIL900 Va22 E8501 PHP85 Oh43  \n",
       "0         G      .    .     G     G    G  \n",
       "1         T      .    G     T     T    G  \n",
       "2         G      G    G     G     G    G  \n",
       "3         G      G    A     G     G    A  \n",
       "4         A      A    A     A     A    A  \n",
       "...     ...    ...  ...   ...   ...  ...  \n",
       "995780    C      C    C     C     C    C  \n",
       "995781    .      A    A     A     A    A  \n",
       "995782    .      G    G     G     G    G  \n",
       "995783    .      C    C     C     C    C  \n",
       "995784    .      .    .     .     G    A  \n",
       "\n",
       "[995785 rows x 952 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have 995,785 rows × 952 columns. I need roughly 16Gb to simply load the matrix. To do actual data wrangling, I need way more memory 😿"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First 10 columns are metadata. The rest correspond to cultivars.\n",
    "\n",
    "The column names for cultivars is found in `B73_plus_RTAs_snp_matrix_imputed/widiv_942g_979873SNPs_imputed_filteredGenos_withRTA_AGPv4.hmp.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(snp_matrix.dtypes.values)\n",
    "chromosomes = np.unique(snp_matrix['chromosome'].values.astype(str))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 9735 different chromosomes.\n",
      "\n",
      " ['1' '10' '2' '3' '4' '5' '6' '7' '8' '9' 'B73V4_ctg10' 'B73V4_ctg102'\n",
      " 'B73V4_ctg103' 'B73V4_ctg104' 'B73V4_ctg105' 'B73V4_ctg107' 'B73V4_ctg11'\n",
      " 'B73V4_ctg115' 'B73V4_ctg123' 'B73V4_ctg129' 'B73V4_ctg134' 'B73V4_ctg14'\n",
      " 'B73V4_ctg140' 'B73V4_ctg141' 'B73V4_ctg142' 'B73V4_ctg147'\n",
      " 'B73V4_ctg150' 'B73V4_ctg152' 'B73V4_ctg153' 'B73V4_ctg154'\n",
      " 'B73V4_ctg165' 'B73V4_ctg166' 'B73V4_ctg167' 'B73V4_ctg168'\n",
      " 'B73V4_ctg170' 'B73V4_ctg171' 'B73V4_ctg173' 'B73V4_ctg178'\n",
      " 'B73V4_ctg181' 'B73V4_ctg182' 'B73V4_ctg188' 'B73V4_ctg189'\n",
      " 'B73V4_ctg190' 'B73V4_ctg193' 'B73V4_ctg2' 'B73V4_ctg203' 'B73V4_ctg205'\n",
      " 'B73V4_ctg207' 'B73V4_ctg23' 'B73V4_ctg24' 'B73V4_ctg247' 'B73V4_ctg248'\n",
      " 'B73V4_ctg26' 'B73V4_ctg28' 'B73V4_ctg3' 'B73V4_ctg31' 'B73V4_ctg35'\n",
      " 'B73V4_ctg4' 'B73V4_ctg41' 'B73V4_ctg44' 'B73V4_ctg48' 'B73V4_ctg5'\n",
      " 'B73V4_ctg50' 'B73V4_ctg51' 'B73V4_ctg52' 'B73V4_ctg53' 'B73V4_ctg56'\n",
      " 'B73V4_ctg58' 'B73V4_ctg6' 'B73V4_ctg61' 'B73V4_ctg65' 'B73V4_ctg66'\n",
      " 'B73V4_ctg68' 'B73V4_ctg72' 'B73V4_ctg73' 'B73V4_ctg74' 'B73V4_ctg75'\n",
      " 'B73V4_ctg77' 'B73V4_ctg88' 'B73V4_ctg89' 'B73V4_ctg91' 'B73V4_ctg95'\n",
      " 'B73V4_ctg98' 'locus_DN100007_c1_g1_i3_length2428'\n",
      " 'locus_DN100009_c1_g1_i2_length802' 'locus_DN100010_c5_g2_i2_length829'\n",
      " 'locus_DN100013_c1_g2_i1_length865' 'locus_DN100021_c3_g1_i1_length880'\n",
      " 'locus_DN100032_c0_g1_i1_length578' 'locus_DN100041_c0_g1_i9_length1220'\n",
      " 'locus_DN100044_c2_g2_i3_length1151' 'locus_DN100049_c0_g1_i5_length1040'\n",
      " 'locus_DN100070_c0_g1_i4_length1176' 'locus_DN100089_c0_g1_i3_length621'\n",
      " 'locus_DN100107_c0_g1_i2_length770' 'locus_DN100111_c1_g2_i1_length544'\n",
      " 'locus_DN100115_c1_g1_i2_length1843' 'locus_DN100115_c2_g1_i2_length996'\n",
      " 'locus_DN100124_c0_g1_i15_length914' 'locus_DN100127_c0_g1_i1_length1036'\n",
      " 'locus_DN100128_c0_g2_i4_length655' 'locus_DN100136_c1_g1_i7_length1498'\n",
      " 'locus_DN100137_c0_g1_i1_length663' 'locus_DN100145_c0_g1_i1_length560'\n",
      " 'locus_DN100145_c0_g2_i8_length1248' 'locus_DN100158_c0_g1_i1_length678'] \n",
      "\n",
      " ['locus_DN125388_c0_g3_i10_length2353'\n",
      " 'locus_DN125391_c0_g1_i7_length1585' 'locus_DN125393_c1_g1_i1_length783'\n",
      " 'locus_DN125403_c2_g1_i1_length1382' 'locus_DN125403_c2_g2_i2_length1186'\n",
      " 'locus_DN125408_c2_g2_i6_length2062' 'locus_DN125409_c0_g1_i2_length1388'\n",
      " 'locus_DN125412_c0_g2_i2_length1059' 'locus_DN125418_c1_g1_i1_length579'\n",
      " 'locus_DN125419_c0_g1_i2_length1390' 'locus_DN125429_c1_g1_i3_length720'\n",
      " 'locus_DN125429_c1_g2_i2_length1406' 'locus_DN125431_c0_g1_i2_length1693'\n",
      " 'locus_DN125437_c0_g1_i2_length838' 'locus_DN125448_c0_g1_i1_length737'\n",
      " 'locus_DN125448_c0_g3_i2_length1654'\n",
      " 'locus_DN125452_c2_g1_i10_length2416' 'locus_DN125453_c2_g1_i1_length632'\n",
      " 'locus_DN125453_c2_g2_i1_length703' 'locus_DN125455_c1_g1_i1_length603']\n"
     ]
    }
   ],
   "source": [
    "print('There are', len(chromosomes), 'different chromosomes.')\n",
    "print('\\n', chromosomes[:106], '\\n\\n', chromosomes[5000:5020])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "downIndex = np.array([])\n",
    "for i in range(1,11):\n",
    "    chromo = snp_matrix[snp_matrix['chromosome'] == '{}'.format(i)]\n",
    "    downIndex = np.concatenate((downIndex, chromo.index.values))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>Allele Counts</th>\n",
       "      <th>Missing Data</th>\n",
       "      <th>Total Calls</th>\n",
       "      <th>Percent A</th>\n",
       "      <th>Percent T</th>\n",
       "      <th>Percent C</th>\n",
       "      <th>Percent G</th>\n",
       "      <th>...</th>\n",
       "      <th>NC328</th>\n",
       "      <th>NC326</th>\n",
       "      <th>PHV53</th>\n",
       "      <th>DKIBC2</th>\n",
       "      <th>A641</th>\n",
       "      <th>WIL900</th>\n",
       "      <th>Va22</th>\n",
       "      <th>E8501</th>\n",
       "      <th>PHP85</th>\n",
       "      <th>Oh43</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>82131.0</th>\n",
       "      <td>1</td>\n",
       "      <td>100085818</td>\n",
       "      <td>C</td>\n",
       "      <td>A(0)T(21)C(508)G(0)</td>\n",
       "      <td>430</td>\n",
       "      <td>529</td>\n",
       "      <td>0</td>\n",
       "      <td>3.96975425330813</td>\n",
       "      <td>96.0302457466919</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82132.0</th>\n",
       "      <td>1</td>\n",
       "      <td>100086117</td>\n",
       "      <td>G</td>\n",
       "      <td>A(30)T(0)C(0)G(275)</td>\n",
       "      <td>654</td>\n",
       "      <td>305</td>\n",
       "      <td>9.83606557377049</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>90.1639344262295</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82133.0</th>\n",
       "      <td>1</td>\n",
       "      <td>100086147</td>\n",
       "      <td>C</td>\n",
       "      <td>A(0)T(0)C(299)G(33)</td>\n",
       "      <td>627</td>\n",
       "      <td>332</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>90.0602409638554</td>\n",
       "      <td>9.93975903614458</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82134.0</th>\n",
       "      <td>1</td>\n",
       "      <td>100086204</td>\n",
       "      <td>G</td>\n",
       "      <td>A(53)T(0)C(0)G(304)</td>\n",
       "      <td>602</td>\n",
       "      <td>357</td>\n",
       "      <td>14.8459383753501</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>85.1540616246499</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82135.0</th>\n",
       "      <td>1</td>\n",
       "      <td>100086286</td>\n",
       "      <td>C</td>\n",
       "      <td>A(0)T(0)C(434)G(54)</td>\n",
       "      <td>471</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>88.9344262295082</td>\n",
       "      <td>11.0655737704918</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995780.0</th>\n",
       "      <td>10</td>\n",
       "      <td>99998392</td>\n",
       "      <td>C</td>\n",
       "      <td>A(0)T(5)C(954)G(0)</td>\n",
       "      <td>0</td>\n",
       "      <td>959</td>\n",
       "      <td>0</td>\n",
       "      <td>0.521376433785193</td>\n",
       "      <td>99.4786235662148</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995781.0</th>\n",
       "      <td>10</td>\n",
       "      <td>99998427</td>\n",
       "      <td>A</td>\n",
       "      <td>A(855)T(0)C(30)G(0)</td>\n",
       "      <td>74</td>\n",
       "      <td>885</td>\n",
       "      <td>96.6101694915254</td>\n",
       "      <td>0</td>\n",
       "      <td>3.38983050847458</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995782.0</th>\n",
       "      <td>10</td>\n",
       "      <td>99998428</td>\n",
       "      <td>G</td>\n",
       "      <td>A(7)T(0)C(0)G(857)</td>\n",
       "      <td>95</td>\n",
       "      <td>864</td>\n",
       "      <td>0.810185185185185</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>99.1898148148148</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995783.0</th>\n",
       "      <td>10</td>\n",
       "      <td>99998430</td>\n",
       "      <td>C</td>\n",
       "      <td>A(0)T(0)C(843)G(5)</td>\n",
       "      <td>111</td>\n",
       "      <td>848</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>99.4103773584906</td>\n",
       "      <td>0.589622641509434</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995784.0</th>\n",
       "      <td>10</td>\n",
       "      <td>99998441</td>\n",
       "      <td>G</td>\n",
       "      <td>A(437)T(0)C(0)G(136)</td>\n",
       "      <td>386</td>\n",
       "      <td>573</td>\n",
       "      <td>76.2652705061082</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23.7347294938918</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>913654 rows × 952 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chromosome   position reference_allele         Allele Counts  \\\n",
       "82131.0           1  100085818                C   A(0)T(21)C(508)G(0)   \n",
       "82132.0           1  100086117                G   A(30)T(0)C(0)G(275)   \n",
       "82133.0           1  100086147                C   A(0)T(0)C(299)G(33)   \n",
       "82134.0           1  100086204                G   A(53)T(0)C(0)G(304)   \n",
       "82135.0           1  100086286                C   A(0)T(0)C(434)G(54)   \n",
       "...             ...        ...              ...                   ...   \n",
       "995780.0         10   99998392                C    A(0)T(5)C(954)G(0)   \n",
       "995781.0         10   99998427                A   A(855)T(0)C(30)G(0)   \n",
       "995782.0         10   99998428                G    A(7)T(0)C(0)G(857)   \n",
       "995783.0         10   99998430                C    A(0)T(0)C(843)G(5)   \n",
       "995784.0         10   99998441                G  A(437)T(0)C(0)G(136)   \n",
       "\n",
       "          Missing Data  Total Calls          Percent A          Percent T  \\\n",
       "82131.0            430          529                  0   3.96975425330813   \n",
       "82132.0            654          305   9.83606557377049                  0   \n",
       "82133.0            627          332                  0                  0   \n",
       "82134.0            602          357   14.8459383753501                  0   \n",
       "82135.0            471          488                  0                  0   \n",
       "...                ...          ...                ...                ...   \n",
       "995780.0             0          959                  0  0.521376433785193   \n",
       "995781.0            74          885   96.6101694915254                  0   \n",
       "995782.0            95          864  0.810185185185185                  0   \n",
       "995783.0           111          848                  0                  0   \n",
       "995784.0           386          573   76.2652705061082                  0   \n",
       "\n",
       "                 Percent C          Percent G  ... NC328 NC326 PHV53 DKIBC2  \\\n",
       "82131.0   96.0302457466919                  0  ...     .     C     C      C   \n",
       "82132.0                  0   90.1639344262295  ...     .     G     G      A   \n",
       "82133.0   90.0602409638554   9.93975903614458  ...     C     C     C      G   \n",
       "82134.0                  0   85.1540616246499  ...     G     G     G      G   \n",
       "82135.0   88.9344262295082   11.0655737704918  ...     C     C     C      G   \n",
       "...                    ...                ...  ...   ...   ...   ...    ...   \n",
       "995780.0  99.4786235662148                  0  ...     C     C     C      C   \n",
       "995781.0  3.38983050847458                  0  ...     A     A     A      A   \n",
       "995782.0                 0   99.1898148148148  ...     G     G     G      G   \n",
       "995783.0  99.4103773584906  0.589622641509434  ...     C     C     C      C   \n",
       "995784.0                 0   23.7347294938918  ...     .     G     G      A   \n",
       "\n",
       "         A641 WIL900 Va22 E8501 PHP85 Oh43  \n",
       "82131.0     .      C    .     C     .    C  \n",
       "82132.0     G      G    G     G     .    G  \n",
       "82133.0     C      C    C     C     .    C  \n",
       "82134.0     A      G    G     G     .    G  \n",
       "82135.0     C      C    C     C     .    C  \n",
       "...       ...    ...  ...   ...   ...  ...  \n",
       "995780.0    C      C    C     C     C    C  \n",
       "995781.0    .      A    A     A     A    A  \n",
       "995782.0    .      G    G     G     G    G  \n",
       "995783.0    .      C    C     C     C    C  \n",
       "995784.0    .      .    .     .     G    A  \n",
       "\n",
       "[913654 rows x 952 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_matrix.loc[downIndex]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "913654"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "995784 - 82130"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "b73v4_ctg10 = snp_matrix[snp_matrix['chromosome'] == '1']\n",
    "b73v4_ctg107 = snp_matrix[snp_matrix['chromosome'] == 'B73V4_ctg107']\n",
    "b73v4_ctg105 = snp_matrix[snp_matrix['chromosome'] == 'B73V4_ctg105']\n",
    "b73v4_ctg104 = snp_matrix[snp_matrix['chromosome'] == 'B73V4_ctg104']\n",
    "b73v4_ctg10.index.values\n",
    "downIndex = np.concatenate((b73v4_ctg10.index.values, b73v4_ctg104.index.values, b73v4_ctg105.index.values, b73v4_ctg107.index.values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have\t 184 \tSNPs in chromosome b73v4_ctg10\n",
      "We have\t 9 \tSNPs in chromosome b73v4_ctg104\n",
      "We have\t 51 \tSNPs in chromosome b73v4_ctg105\n",
      "We have\t 277 \tSNPs in chromosome b73v4_ctg107\n",
      "TOTAL\t 521\n",
      "Between 1 and 635\n"
     ]
    }
   ],
   "source": [
    "print('We have\\t', len(b73v4_ctg10), '\\tSNPs in chromosome b73v4_ctg10')\n",
    "print('We have\\t', len(b73v4_ctg104), '\\tSNPs in chromosome b73v4_ctg104')\n",
    "print('We have\\t', len(b73v4_ctg105), '\\tSNPs in chromosome b73v4_ctg105')\n",
    "print('We have\\t', len(b73v4_ctg107), '\\tSNPs in chromosome b73v4_ctg107')\n",
    "print('TOTAL\\t', len(downIndex))\n",
    "print('Between', downIndex.min(), 'and', downIndex.max())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "downsample = snp_matrix.loc[downIndex]\n",
    "downsample.to_csv(dst+'snp_matrix_downsample.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "downsample = pd.read_csv(dst+'snp_matrix_downsample.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A total of 512 SNPs have different Missing_Data as reported.\n",
      "Largest discrepancy:  17 [ at row 223 ]\n"
     ]
    }
   ],
   "source": [
    "missing_data = downsample.apply(numMissing, axis=1).values\n",
    "discrepancy_missing = downsample['Missing Data'].values - missing_data\n",
    "print('A total of', len(discrepancy_missing != 0), 'SNPs have different Missing_Data as reported.')\n",
    "print('Largest discrepancy: ', discrepancy_missing.max(), '[ at row', discrepancy_missing.argmax(),']' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean total:\t 959.0\n",
      "Var total:\t 0.0\n"
     ]
    }
   ],
   "source": [
    "total_data = downsample.apply(numData, axis=1).values\n",
    "print('Mean total:\\t', total_data.mean())\n",
    "print('Var total:\\t', total_data.var())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The given data assumes **959** total calls (cultivars?), but we only have **942** available cultivars. Thus the `Missing Data` column has **wrong** values."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's explore the expression data!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ejam/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>position_left</th>\n",
       "      <th>position_right</th>\n",
       "      <th>LH128</th>\n",
       "      <th>DKMBZA</th>\n",
       "      <th>CQ806</th>\n",
       "      <th>DKF274</th>\n",
       "      <th>Ill.Hy</th>\n",
       "      <th>...</th>\n",
       "      <th>NC328</th>\n",
       "      <th>NC326</th>\n",
       "      <th>PHV53</th>\n",
       "      <th>DKIBC2</th>\n",
       "      <th>A641</th>\n",
       "      <th>WIL900</th>\n",
       "      <th>Va22</th>\n",
       "      <th>E8501</th>\n",
       "      <th>PHP85</th>\n",
       "      <th>Oh43</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d012719</td>\n",
       "      <td>8</td>\n",
       "      <td>gene</td>\n",
       "      <td>179164454</td>\n",
       "      <td>179168169</td>\n",
       "      <td>5.825000</td>\n",
       "      <td>4.845710</td>\n",
       "      <td>6.323400</td>\n",
       "      <td>6.999010</td>\n",
       "      <td>6.661100</td>\n",
       "      <td>...</td>\n",
       "      <td>15.154700</td>\n",
       "      <td>12.870700</td>\n",
       "      <td>8.81507</td>\n",
       "      <td>13.77870</td>\n",
       "      <td>8.31790</td>\n",
       "      <td>12.78320</td>\n",
       "      <td>4.94950</td>\n",
       "      <td>9.39900</td>\n",
       "      <td>10.41090</td>\n",
       "      <td>15.840500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Zm00001d024742</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>85863323</td>\n",
       "      <td>85863746</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Zm00001d007137</td>\n",
       "      <td>2</td>\n",
       "      <td>tRNA_gene</td>\n",
       "      <td>223153553</td>\n",
       "      <td>223153627</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Zm00001d025653</td>\n",
       "      <td>10</td>\n",
       "      <td>gene</td>\n",
       "      <td>125107973</td>\n",
       "      <td>125113328</td>\n",
       "      <td>9.350780</td>\n",
       "      <td>9.954800</td>\n",
       "      <td>7.735490</td>\n",
       "      <td>10.360900</td>\n",
       "      <td>9.072890</td>\n",
       "      <td>...</td>\n",
       "      <td>4.685600</td>\n",
       "      <td>3.870230</td>\n",
       "      <td>5.27849</td>\n",
       "      <td>5.17829</td>\n",
       "      <td>3.80291</td>\n",
       "      <td>6.44675</td>\n",
       "      <td>5.61858</td>\n",
       "      <td>3.98055</td>\n",
       "      <td>4.17545</td>\n",
       "      <td>4.500850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Zm00001d036003</td>\n",
       "      <td>6</td>\n",
       "      <td>gene</td>\n",
       "      <td>65322472</td>\n",
       "      <td>65323171</td>\n",
       "      <td>8.955580</td>\n",
       "      <td>0.333921</td>\n",
       "      <td>5.445820</td>\n",
       "      <td>0.332438</td>\n",
       "      <td>1.232330</td>\n",
       "      <td>...</td>\n",
       "      <td>0.232934</td>\n",
       "      <td>0.721774</td>\n",
       "      <td>2.95424</td>\n",
       "      <td>8.44716</td>\n",
       "      <td>12.33680</td>\n",
       "      <td>1.19629</td>\n",
       "      <td>1.71452</td>\n",
       "      <td>7.01768</td>\n",
       "      <td>11.75760</td>\n",
       "      <td>0.793672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44254</th>\n",
       "      <td>Zm00001d028653</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>41833737</td>\n",
       "      <td>41838204</td>\n",
       "      <td>368.777000</td>\n",
       "      <td>301.349000</td>\n",
       "      <td>412.319000</td>\n",
       "      <td>310.820000</td>\n",
       "      <td>278.737000</td>\n",
       "      <td>...</td>\n",
       "      <td>172.293000</td>\n",
       "      <td>186.552000</td>\n",
       "      <td>147.63300</td>\n",
       "      <td>165.60800</td>\n",
       "      <td>147.58300</td>\n",
       "      <td>244.82100</td>\n",
       "      <td>171.15400</td>\n",
       "      <td>219.07400</td>\n",
       "      <td>147.01700</td>\n",
       "      <td>165.612000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44255</th>\n",
       "      <td>Zm00001d009511</td>\n",
       "      <td>8</td>\n",
       "      <td>gene</td>\n",
       "      <td>68418436</td>\n",
       "      <td>68421217</td>\n",
       "      <td>6.003590</td>\n",
       "      <td>5.297160</td>\n",
       "      <td>10.421700</td>\n",
       "      <td>6.304130</td>\n",
       "      <td>4.856300</td>\n",
       "      <td>...</td>\n",
       "      <td>4.120700</td>\n",
       "      <td>4.670860</td>\n",
       "      <td>3.47752</td>\n",
       "      <td>4.92093</td>\n",
       "      <td>6.26904</td>\n",
       "      <td>4.95873</td>\n",
       "      <td>7.25625</td>\n",
       "      <td>5.66864</td>\n",
       "      <td>4.28185</td>\n",
       "      <td>5.765310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44256</th>\n",
       "      <td>Zm00001d032828</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>239455289</td>\n",
       "      <td>239464371</td>\n",
       "      <td>151.479000</td>\n",
       "      <td>46.479600</td>\n",
       "      <td>65.075600</td>\n",
       "      <td>70.169900</td>\n",
       "      <td>59.555600</td>\n",
       "      <td>...</td>\n",
       "      <td>100.249000</td>\n",
       "      <td>102.226000</td>\n",
       "      <td>100.29500</td>\n",
       "      <td>64.90400</td>\n",
       "      <td>91.17440</td>\n",
       "      <td>101.80300</td>\n",
       "      <td>108.58200</td>\n",
       "      <td>63.72140</td>\n",
       "      <td>95.96760</td>\n",
       "      <td>56.404900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44257</th>\n",
       "      <td>Zm00001d002031</td>\n",
       "      <td>2</td>\n",
       "      <td>gene</td>\n",
       "      <td>4673452</td>\n",
       "      <td>4674247</td>\n",
       "      <td>0.237774</td>\n",
       "      <td>0.386886</td>\n",
       "      <td>0.179111</td>\n",
       "      <td>0.415566</td>\n",
       "      <td>0.192793</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1.51552</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>2.555730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44258</th>\n",
       "      <td>Zm00001d011196</td>\n",
       "      <td>8</td>\n",
       "      <td>tRNA_gene</td>\n",
       "      <td>142721025</td>\n",
       "      <td>142721097</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>44259 rows × 947 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene chromosome feature_type  position_left  position_right  \\\n",
       "0      Zm00001d012719          8         gene      179164454       179168169   \n",
       "1      Zm00001d024742         10         gene       85863323        85863746   \n",
       "2      Zm00001d007137          2    tRNA_gene      223153553       223153627   \n",
       "3      Zm00001d025653         10         gene      125107973       125113328   \n",
       "4      Zm00001d036003          6         gene       65322472        65323171   \n",
       "...               ...        ...          ...            ...             ...   \n",
       "44254  Zm00001d028653          1         gene       41833737        41838204   \n",
       "44255  Zm00001d009511          8         gene       68418436        68421217   \n",
       "44256  Zm00001d032828          1         gene      239455289       239464371   \n",
       "44257  Zm00001d002031          2         gene        4673452         4674247   \n",
       "44258  Zm00001d011196          8    tRNA_gene      142721025       142721097   \n",
       "\n",
       "            LH128      DKMBZA       CQ806      DKF274      Ill.Hy  ...  \\\n",
       "0        5.825000    4.845710    6.323400    6.999010    6.661100  ...   \n",
       "1        0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "2        0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "3        9.350780    9.954800    7.735490   10.360900    9.072890  ...   \n",
       "4        8.955580    0.333921    5.445820    0.332438    1.232330  ...   \n",
       "...           ...         ...         ...         ...         ...  ...   \n",
       "44254  368.777000  301.349000  412.319000  310.820000  278.737000  ...   \n",
       "44255    6.003590    5.297160   10.421700    6.304130    4.856300  ...   \n",
       "44256  151.479000   46.479600   65.075600   70.169900   59.555600  ...   \n",
       "44257    0.237774    0.386886    0.179111    0.415566    0.192793  ...   \n",
       "44258    0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "\n",
       "            NC328       NC326      PHV53     DKIBC2       A641     WIL900  \\\n",
       "0       15.154700   12.870700    8.81507   13.77870    8.31790   12.78320   \n",
       "1        0.000000    0.000000    0.00000    0.00000    0.00000    0.00000   \n",
       "2        0.000000    0.000000    0.00000    0.00000    0.00000    0.00000   \n",
       "3        4.685600    3.870230    5.27849    5.17829    3.80291    6.44675   \n",
       "4        0.232934    0.721774    2.95424    8.44716   12.33680    1.19629   \n",
       "...           ...         ...        ...        ...        ...        ...   \n",
       "44254  172.293000  186.552000  147.63300  165.60800  147.58300  244.82100   \n",
       "44255    4.120700    4.670860    3.47752    4.92093    6.26904    4.95873   \n",
       "44256  100.249000  102.226000  100.29500   64.90400   91.17440  101.80300   \n",
       "44257    0.000000    0.000000    0.00000    0.00000    0.00000    1.51552   \n",
       "44258    0.000000    0.000000    0.00000    0.00000    0.00000    0.00000   \n",
       "\n",
       "            Va22      E8501      PHP85        Oh43  \n",
       "0        4.94950    9.39900   10.41090   15.840500  \n",
       "1        0.00000    0.00000    0.00000    0.000000  \n",
       "2        0.00000    0.00000    0.00000    0.000000  \n",
       "3        5.61858    3.98055    4.17545    4.500850  \n",
       "4        1.71452    7.01768   11.75760    0.793672  \n",
       "...          ...        ...        ...         ...  \n",
       "44254  171.15400  219.07400  147.01700  165.612000  \n",
       "44255    7.25625    5.66864    4.28185    5.765310  \n",
       "44256  108.58200   63.72140   95.96760   56.404900  \n",
       "44257    0.00000    0.00000    0.00000    2.555730  \n",
       "44258    0.00000    0.00000    0.00000    0.000000  \n",
       "\n",
       "[44259 rows x 947 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_w_features = pd.read_table(src+file_gen)\n",
    "genes_w_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 106 different chromosomes.\n",
      "['1' '10' '2' '3' '4' '5' '6' '7' '8' '9' 'B73V4_ctg10' 'B73V4_ctg100'\n",
      " 'B73V4_ctg102' 'B73V4_ctg103' 'B73V4_ctg104' 'B73V4_ctg105'\n",
      " 'B73V4_ctg107' 'B73V4_ctg11' 'B73V4_ctg110' 'B73V4_ctg115' 'B73V4_ctg117'\n",
      " 'B73V4_ctg119' 'B73V4_ctg12' 'B73V4_ctg120' 'B73V4_ctg121' 'B73V4_ctg123'\n",
      " 'B73V4_ctg125' 'B73V4_ctg129' 'B73V4_ctg13' 'B73V4_ctg133' 'B73V4_ctg134'\n",
      " 'B73V4_ctg14' 'B73V4_ctg140' 'B73V4_ctg141' 'B73V4_ctg142' 'B73V4_ctg144'\n",
      " 'B73V4_ctg147' 'B73V4_ctg150' 'B73V4_ctg151' 'B73V4_ctg170'\n",
      " 'B73V4_ctg173' 'B73V4_ctg178' 'B73V4_ctg18' 'B73V4_ctg180' 'B73V4_ctg181'\n",
      " 'B73V4_ctg182' 'B73V4_ctg188' 'B73V4_ctg189' 'B73V4_ctg190'\n",
      " 'B73V4_ctg193' 'B73V4_ctg2' 'B73V4_ctg20' 'B73V4_ctg205' 'B73V4_ctg206'\n",
      " 'B73V4_ctg208' 'B73V4_ctg219' 'B73V4_ctg23' 'B73V4_ctg24' 'B73V4_ctg242'\n",
      " 'B73V4_ctg245' 'B73V4_ctg247' 'B73V4_ctg248' 'B73V4_ctg250'\n",
      " 'B73V4_ctg253' 'B73V4_ctg26' 'B73V4_ctg28' 'B73V4_ctg29' 'B73V4_ctg3'\n",
      " 'B73V4_ctg31' 'B73V4_ctg35' 'B73V4_ctg4' 'B73V4_ctg40' 'B73V4_ctg41'\n",
      " 'B73V4_ctg42' 'B73V4_ctg44' 'B73V4_ctg46' 'B73V4_ctg48' 'B73V4_ctg49'\n",
      " 'B73V4_ctg5' 'B73V4_ctg50' 'B73V4_ctg51' 'B73V4_ctg52' 'B73V4_ctg53'\n",
      " 'B73V4_ctg54' 'B73V4_ctg56' 'B73V4_ctg58' 'B73V4_ctg6' 'B73V4_ctg61'\n",
      " 'B73V4_ctg65' 'B73V4_ctg66' 'B73V4_ctg68' 'B73V4_ctg69' 'B73V4_ctg72'\n",
      " 'B73V4_ctg73' 'B73V4_ctg74' 'B73V4_ctg75' 'B73V4_ctg77' 'B73V4_ctg8'\n",
      " 'B73V4_ctg83' 'B73V4_ctg88' 'B73V4_ctg89' 'B73V4_ctg91' 'B73V4_ctg92'\n",
      " 'B73V4_ctg95' 'B73V4_ctg97' 'B73V4_ctg98']\n"
     ]
    }
   ],
   "source": [
    "bar = np.unique(genes_w_features['chromosome'].values.astype(str))\n",
    "print('There are', len(bar), 'different chromosomes.')\n",
    "\n",
    "print(bar)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have 44259 rows × 947 columns. The set of chromosomes in both data files **are different!!!** **Neither** is a subset of the other. Observe that `'B73V4_ctg100'` shows up only in `genes_w_features`. Both data sets contain 942 cultivars."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have\t 12 \tSNPs in chromosome b73v4_ctg10\n",
      "We have\t 3 \tSNPs in chromosome b73v4_ctg104\n",
      "We have\t 1 \tSNPs in chromosome b73v4_ctg105\n",
      "We have\t 8 \tSNPs in chromosome b73v4_ctg107\n",
      "TOTAL\t 24\n",
      "Between 459 and 41134\n"
     ]
    }
   ],
   "source": [
    "b73v4_ctg10 = genes_w_features[genes_w_features['chromosome'] == 'B73V4_ctg10']\n",
    "b73v4_ctg107 = genes_w_features[genes_w_features['chromosome'] == 'B73V4_ctg107']\n",
    "b73v4_ctg105 = genes_w_features[genes_w_features['chromosome'] == 'B73V4_ctg105']\n",
    "b73v4_ctg104 = genes_w_features[genes_w_features['chromosome'] == 'B73V4_ctg104']\n",
    "b73v4_ctg10.index.values\n",
    "downIndex = np.concatenate((b73v4_ctg10.index.values, b73v4_ctg104.index.values, b73v4_ctg105.index.values, b73v4_ctg107.index.values))\n",
    "\n",
    "print('We have\\t', len(b73v4_ctg10), '\\tSNPs in chromosome b73v4_ctg10')\n",
    "print('We have\\t', len(b73v4_ctg104), '\\tSNPs in chromosome b73v4_ctg104')\n",
    "print('We have\\t', len(b73v4_ctg105), '\\tSNPs in chromosome b73v4_ctg105')\n",
    "print('We have\\t', len(b73v4_ctg107), '\\tSNPs in chromosome b73v4_ctg107')\n",
    "print('TOTAL\\t', len(downIndex))\n",
    "print('Between', downIndex.min(), 'and', downIndex.max())\n",
    "\n",
    "downsample = genes_w_features.loc[downIndex]\n",
    "downsample.to_csv(dst+'genes_w_features_downsample.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
